{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-21 15:55:28.049218: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2025-05-21 15:55:28.168565: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory\n",
      "2025-05-21 15:55:28.168592: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2025-05-21 15:55:30.909960: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory\n",
      "2025-05-21 15:55:30.910061: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory\n",
      "2025-05-21 15:55:30.910075: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%cd -q ..\n",
    "\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from evoaug_tf import evoaug, augment\n",
    "from src.diff_expression_model import get_model, get_siamese_model, post_hoc_conjoining, get_auroc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now we start from this data pickle as I'm not aware how Helder did DE analysis and generated the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppath = \"data/dataset_solid_chrome.csv\"\n",
    "data = pd.read_csv(ppath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneID</th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>seqs</th>\n",
       "      <th>ohs</th>\n",
       "      <th>rcohs</th>\n",
       "      <th>ohsDuo</th>\n",
       "      <th>in_original_balanced</th>\n",
       "      <th>set</th>\n",
       "      <th>npshap-single</th>\n",
       "      <th>npshap-posthoc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0 0 0 1]\\n [1 0 0 0]\\n [0 0 1 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[1 0 0 0]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [1 0 0 0]\\n  [0 0 1 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[array([[[-2.94115947e-04],\\n        [ 1.11911...</td>\n",
       "      <td>[array([[[-2.47588614e-04],\\n        [ 1.33605...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0 0 1 0]\\n [0 0 0 1]\\n [0 1 0 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[0 0 1 0]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[0 0 1 0]\\n  [0 0 0 1]\\n  [0 1 0 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[array([[[ 4.96114546e-04],\\n        [ 6.92353...</td>\n",
       "      <td>[array([[[ 5.19316818e-04],\\n        [ 3.21869...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1 0 0 0]\\n [0 0 1 0]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[[1 0 0 0]\\n  [0 0 1 0]\\n  [0 0 0 1]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[array([[[-3.02427739e-04],\\n        [-1.08283...</td>\n",
       "      <td>[array([[[-3.14956094e-04],\\n        [-8.34961...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0 1 0 0]\\n [1 0 0 0]\\n [0 1 0 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[1 0 0 0]\\n [0 1 0 0]\\n [0 0 1 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[0 1 0 0]\\n  [1 0 0 0]\\n  [0 1 0 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[array([[[ 2.54294393e-03],\\n        [ 6.39127...</td>\n",
       "      <td>[array([[[ 2.08252599e-03],\\n        [ 1.86260...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 0 1]\\n [0 0 1 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[0 0 0 1]\\n [1 0 0 0]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 1 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[array([[[-5.23597009e-05],\\n        [ 4.53725...</td>\n",
       "      <td>[array([[[ 5.18305111e-04],\\n        [ 5.07344...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1 0 0 0]\\n [1 0 0 0]\\n [0 0 1 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[0 1 0 0]\\n [0 0 1 0]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[1 0 0 0]\\n  [1 0 0 0]\\n  [0 0 1 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[array([[[-1.85048248e-05],\\n        [ 5.31432...</td>\n",
       "      <td>[array([[[ 3.42493469e-04],\\n        [ 3.91582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 0 1]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 0 1]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[array([[[ 9.53665585e-05],\\n        [ 3.28118...</td>\n",
       "      <td>[array([[[-5.87131144e-05],\\n        [ 6.68894...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0 0 0 1]\\n [1 0 0 0]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[1 0 0 0]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [1 0 0 0]\\n  [0 0 0 1]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[array([[[ 6.18251797e-04],\\n        [ 6.60565...</td>\n",
       "      <td>[array([[[ 4.48545878e-04],\\n        [ 6.36329...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 0 1]\\n [0 0 1 0]\\n ...\\n [0 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [0 0 1 0]\\n  [0 0 1 0]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[array([[[-2.15267210e-04],\\n        [ 3.76152...</td>\n",
       "      <td>[array([[[-1.03627557e-04],\\n        [ 1.56308...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0 0 0 1]\\n [0 0 0 1]\\n [0 0 0 1]\\n ...\\n [0 ...</td>\n",
       "      <td>[[0 0 1 0]\\n [0 0 1 0]\\n [0 1 0 0]\\n ...\\n [1 ...</td>\n",
       "      <td>[[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 0 1]\\n  ...\\n...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[array([[[ 6.95527473e-04],\\n        [ 2.36141...</td>\n",
       "      <td>[array([[[ 4.71143954e-04],\\n        [ 9.89265...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         geneID  Category    GeneFamily  \\\n",
       "0     AT4G27120         0  HOM04D000881   \n",
       "1     AT4G19600         0  HOM04D000740   \n",
       "2     AT3G60880         0  HOM04D003119   \n",
       "3     AT5G06960         0  HOM04D000319   \n",
       "4     AT1G14890         0  HOM04D000273   \n",
       "...         ...       ...           ...   \n",
       "2393  AT5G64230         1  HOM04D003278   \n",
       "2394  AT5G64780         1  HOM04D002552   \n",
       "2395  AT4G30470         1  HOM04D000082   \n",
       "2396  AT3G51895         1  HOM04D000270   \n",
       "2397  AT2G35585         1  HOM04D004118   \n",
       "\n",
       "                                                   seqs  \\\n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                 ...   \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                                    ohs  \\\n",
       "0     [[0 0 0 1]\\n [1 0 0 0]\\n [0 0 1 0]\\n ...\\n [0 ...   \n",
       "1     [[0 0 1 0]\\n [0 0 0 1]\\n [0 1 0 0]\\n ...\\n [0 ...   \n",
       "2     [[1 0 0 0]\\n [0 0 1 0]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "3     [[0 1 0 0]\\n [1 0 0 0]\\n [0 1 0 0]\\n ...\\n [0 ...   \n",
       "4     [[0 0 0 1]\\n [0 0 0 1]\\n [0 0 1 0]\\n ...\\n [1 ...   \n",
       "...                                                 ...   \n",
       "2393  [[1 0 0 0]\\n [1 0 0 0]\\n [0 0 1 0]\\n ...\\n [1 ...   \n",
       "2394  [[0 0 0 1]\\n [0 0 0 1]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "2395  [[0 0 0 1]\\n [1 0 0 0]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "2396  [[0 0 0 1]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [0 ...   \n",
       "2397  [[0 0 0 1]\\n [0 0 0 1]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "\n",
       "                                                  rcohs  \\\n",
       "0     [[1 0 0 0]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [0 ...   \n",
       "1     [[0 0 1 0]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [0 ...   \n",
       "2     [[0 0 0 1]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [1 ...   \n",
       "3     [[1 0 0 0]\\n [0 1 0 0]\\n [0 0 1 0]\\n ...\\n [0 ...   \n",
       "4     [[0 0 0 1]\\n [1 0 0 0]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "...                                                 ...   \n",
       "2393  [[0 1 0 0]\\n [0 0 1 0]\\n [0 0 0 1]\\n ...\\n [0 ...   \n",
       "2394  [[0 0 0 1]\\n [0 0 1 0]\\n [1 0 0 0]\\n ...\\n [1 ...   \n",
       "2395  [[1 0 0 0]\\n [0 0 1 0]\\n [0 0 1 0]\\n ...\\n [1 ...   \n",
       "2396  [[0 0 0 1]\\n [0 0 0 1]\\n [0 0 1 0]\\n ...\\n [0 ...   \n",
       "2397  [[0 0 1 0]\\n [0 0 1 0]\\n [0 1 0 0]\\n ...\\n [1 ...   \n",
       "\n",
       "                                                 ohsDuo  in_original_balanced  \\\n",
       "0     [[[0 0 0 1]\\n  [1 0 0 0]\\n  [0 0 1 0]\\n  ...\\n...                  True   \n",
       "1     [[[0 0 1 0]\\n  [0 0 0 1]\\n  [0 1 0 0]\\n  ...\\n...                  True   \n",
       "2     [[[1 0 0 0]\\n  [0 0 1 0]\\n  [0 0 0 1]\\n  ...\\n...                  True   \n",
       "3     [[[0 1 0 0]\\n  [1 0 0 0]\\n  [0 1 0 0]\\n  ...\\n...                  True   \n",
       "4     [[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 1 0]\\n  ...\\n...                  True   \n",
       "...                                                 ...                   ...   \n",
       "2393  [[[1 0 0 0]\\n  [1 0 0 0]\\n  [0 0 1 0]\\n  ...\\n...                  True   \n",
       "2394  [[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 0 1]\\n  ...\\n...                  True   \n",
       "2395  [[[0 0 0 1]\\n  [1 0 0 0]\\n  [0 0 0 1]\\n  ...\\n...                  True   \n",
       "2396  [[[0 0 0 1]\\n  [0 0 1 0]\\n  [0 0 1 0]\\n  ...\\n...                  True   \n",
       "2397  [[[0 0 0 1]\\n  [0 0 0 1]\\n  [0 0 0 1]\\n  ...\\n...                  True   \n",
       "\n",
       "        set                                      npshap-single  \\\n",
       "0     train  [array([[[-2.94115947e-04],\\n        [ 1.11911...   \n",
       "1     valid  [array([[[ 4.96114546e-04],\\n        [ 6.92353...   \n",
       "2     valid  [array([[[-3.02427739e-04],\\n        [-1.08283...   \n",
       "3     train  [array([[[ 2.54294393e-03],\\n        [ 6.39127...   \n",
       "4     valid  [array([[[-5.23597009e-05],\\n        [ 4.53725...   \n",
       "...     ...                                                ...   \n",
       "2393   test  [array([[[-1.85048248e-05],\\n        [ 5.31432...   \n",
       "2394  train  [array([[[ 9.53665585e-05],\\n        [ 3.28118...   \n",
       "2395   test  [array([[[ 6.18251797e-04],\\n        [ 6.60565...   \n",
       "2396  train  [array([[[-2.15267210e-04],\\n        [ 3.76152...   \n",
       "2397  train  [array([[[ 6.95527473e-04],\\n        [ 2.36141...   \n",
       "\n",
       "                                         npshap-posthoc  \n",
       "0     [array([[[-2.47588614e-04],\\n        [ 1.33605...  \n",
       "1     [array([[[ 5.19316818e-04],\\n        [ 3.21869...  \n",
       "2     [array([[[-3.14956094e-04],\\n        [-8.34961...  \n",
       "3     [array([[[ 2.08252599e-03],\\n        [ 1.86260...  \n",
       "4     [array([[[ 5.18305111e-04],\\n        [ 5.07344...  \n",
       "...                                                 ...  \n",
       "2393  [array([[[ 3.42493469e-04],\\n        [ 3.91582...  \n",
       "2394  [array([[[-5.87131144e-05],\\n        [ 6.68894...  \n",
       "2395  [array([[[ 4.48545878e-04],\\n        [ 6.36329...  \n",
       "2396  [array([[[-1.03627557e-04],\\n        [ 1.56308...  \n",
       "2397  [array([[[ 4.71143954e-04],\\n        [ 9.89265...  \n",
       "\n",
       "[2398 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "train    1900\n",
       "valid     257\n",
       "test      241\n",
       "Name: set, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"set\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To show how to subdivide the dataset into train-test split we only take the following columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = data.reset_index()\n",
    "dataset = dataset[[\"geneID\", \"Category\", \"GeneFamily\", \"seqs\"]]\n",
    "dataset.rename(columns={\"geneID\":\"GeneID\", \"Category\":\"Label\", \"seqs\": \"Sequence\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID  Label    GeneFamily  \\\n",
       "0     AT4G27120      0  HOM04D000881   \n",
       "1     AT4G19600      0  HOM04D000740   \n",
       "2     AT3G60880      0  HOM04D003119   \n",
       "3     AT5G06960      0  HOM04D000319   \n",
       "4     AT1G14890      0  HOM04D000273   \n",
       "...         ...    ...           ...   \n",
       "2393  AT5G64230      1  HOM04D003278   \n",
       "2394  AT5G64780      1  HOM04D002552   \n",
       "2395  AT4G30470      1  HOM04D000082   \n",
       "2396  AT3G51895      1  HOM04D000270   \n",
       "2397  AT2G35585      1  HOM04D004118   \n",
       "\n",
       "                                               Sequence  \n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...  \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...  \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...  \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...  \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...  \n",
       "...                                                 ...  \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...  \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...  \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...  \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...  \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...  \n",
       "\n",
       "[2398 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylo augmentation "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the fasta files for each species of which we want to use orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_augmentation_base_path = \"data/phylo_aug/fasta/\"\n",
    "species_fasta_files = {\n",
    "\t\t\t\"aly\":\"aly.tss.fasta\",\n",
    "\t\t\t\"chi\":\"chi.tss.fasta\",\n",
    "\t\t\t\"cru\":\"cru.tss.fasta\"\n",
    "}\n",
    "\n",
    "species_to_include = [\"aly\",\"chi\",\"cru\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsing the fasta files. In this case the label of the original sequence is incorporated in the ids in the fasta file\n",
    "e.g:\n",
    "```\n",
    ">Aa31LG6G5540_1\n",
    "TTGTATAAATATATTAAATT\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: data/phylo_aug/fasta/aly.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/chi.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/cru.tss.fasta done!\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequence_list = []\n",
    "gene_id_list = []\n",
    "label_list = []\n",
    "species_list = []\n",
    "for species in species_to_include:\n",
    "\tspecies_fasta = species_fasta_files[species]\n",
    "\tfasta_path = f\"{phylo_augmentation_base_path}{species_fasta}\"\n",
    "\tfasta_sequences = SeqIO.parse(open(fasta_path),'fasta')\n",
    "\tprint(f\"File: {fasta_path} done!\")\n",
    "\tfor fasta in fasta_sequences:\n",
    "\t\tname, sequence = fasta.id, str(fasta.seq)\n",
    "\t\t\n",
    "\t\t# The label is stored inside the id name (e.g. >ATERI-4G42060_1), split these\n",
    "\t\tgene_id = name.split(\"_\")[0]\n",
    "\t\tlabel = name.split(\"_\")[1]\n",
    "\t\tspecies = species_fasta.split(\".\")[0]\n",
    "\n",
    "\t\tgene_id_list.append(gene_id)\n",
    "\t\tsequence_list.append(sequence)\n",
    "\t\tlabel_list.append(label)\n",
    "\t\tspecies_list.append(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.DataFrame({\"GeneID\": gene_id_list, \"Label\": label_list, \"Sequence\": sequence_list, \"Species\": species_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  \n",
       "...                                                 ...     ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  \n",
       "\n",
       "[2794 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get gene family of the original ortholog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because during the batch time augmentation a original sample will be swapped by one of its orthologs, we need to know from which original sample each ortholog originates from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir_ortholog = \"data/phylo_aug/orthologs/\"\n",
    "ortholog_files = {\n",
    "\t\t\t\"ath-aly\":\"ath_aly.txt\",\n",
    "\t\t\t\"ath-chi\":\"ath_chi.txt\",\n",
    "\t\t\t\"ath-cru\":\"ath_cru.txt\"\n",
    "\t\t\t}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a lookup table to include the orthologs in the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "lookup_table = {}\n",
    "for orthologs in ortholog_files.values():\n",
    "\twith open(f\"{base_dir_ortholog}{orthologs}\", 'r') as file:\n",
    "\t\tfor line in file:\n",
    "\t\t\t# Split each line by tab\n",
    "\t\t\tcolumns = line.strip().split('\\t')\n",
    "\t\t\tif len(columns) >= 2:  # Ensure there are at least two columns\n",
    "\t\t\t\tkey = columns[1]\n",
    "\t\t\t\tvalue = columns[0]\n",
    "\t\t\t\tlookup_table[key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df[\"Ortholog\"] = phylo_aug_df[\"GeneID\"].map(lookup_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260  \n",
       "...                                                 ...     ...        ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970  \n",
       "\n",
       "[2794 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.merge(phylo_aug_df, dataset[[\"GeneID\", \"GeneFamily\"]], left_on=\"Ortholog\", right_on=\"GeneID\", how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily  \n",
       "0     HOM04D003099  \n",
       "1     HOM04D003060  \n",
       "2     HOM04D000135  \n",
       "3     HOM04D001302  \n",
       "4     HOM04D006130  \n",
       "...            ...  \n",
       "2789  HOM04D004133  \n",
       "2790  HOM04D001924  \n",
       "2791  HOM04D006357  \n",
       "2792  HOM04D007279  \n",
       "2793  HOM04D001540  \n",
       "\n",
       "[2794 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df.drop(\"GeneID_y\", axis=1, inplace=True)\n",
    "phylo_aug_df.rename(columns={\"GeneID_x\": \"GeneID\"}, inplace=True)\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One-hot-encode & reverse-complement both dataset as augmented dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import one_hot_encode_series, reverse_complement_series, reverse_complement_sequence\n",
    "dataset[\"One_hot_encoded\"] = one_hot_encode_series(dataset[\"Sequence\"])\n",
    "phylo_aug_df[\"One_hot_encoded\"] = one_hot_encode_series(phylo_aug_df[\"Sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[\"RC_one_hot_encoded\"] = reverse_complement_series(dataset[\"One_hot_encoded\"])\n",
    "phylo_aug_df[\"RC_one_hot_encoded\"] = reverse_complement_series(phylo_aug_df[\"One_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID  Label    GeneFamily  \\\n",
       "0     AT4G27120      0  HOM04D000881   \n",
       "1     AT4G19600      0  HOM04D000740   \n",
       "2     AT3G60880      0  HOM04D003119   \n",
       "3     AT5G06960      0  HOM04D000319   \n",
       "4     AT1G14890      0  HOM04D000273   \n",
       "...         ...    ...           ...   \n",
       "2393  AT5G64230      1  HOM04D003278   \n",
       "2394  AT5G64780      1  HOM04D002552   \n",
       "2395  AT4G30470      1  HOM04D000082   \n",
       "2396  AT3G51895      1  HOM04D000270   \n",
       "2397  AT2G35585      1  HOM04D004118   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                 ...   \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                        One_hot_encoded  \\\n",
       "0     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "1     [[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...   \n",
       "2     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "3     [[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "4     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "...                                                 ...   \n",
       "2393  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "2394  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "2395  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2396  [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "2397  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "1     [[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2     [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "3     [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...  \n",
       "4     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...  \n",
       "...                                                 ...  \n",
       "2393  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...  \n",
       "2394  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2395  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "2396  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...  \n",
       "2397  [[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "\n",
       "[2398 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     HOM04D006130  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...            ...                                                ...   \n",
       "2789  HOM04D004133  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2791  HOM04D006357  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2789  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2791  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2794 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     HOM04D006130  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...            ...                                                ...   \n",
       "2789  HOM04D004133  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2791  HOM04D006357  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2789  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2791  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2794 rows x 8 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df = phylo_aug_df[phylo_aug_df[\"GeneFamily\"].notna()]\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train-test split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import grouped_shuffle_split\n",
    "train_df, validation_test_df = grouped_shuffle_split(dataset, dataset[\"GeneFamily\"], 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_df, test_df  = grouped_shuffle_split(validation_test_df, validation_test_df[\"GeneFamily\"], 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of training set: 1900\n",
      "Length of validation set: 257\n",
      "Length of test set: 241\n"
     ]
    }
   ],
   "source": [
    "print(f\"Length of training set: {train_df.shape[0]}\")\n",
    "print(f\"Length of validation set: {validation_df.shape[0]}\")\n",
    "print(f\"Length of test set: {test_df.shape[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Append train and augmentation dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out the orthologs from samples in test or validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_phylo_aug_df = phylo_aug_df[~phylo_aug_df[\"Ortholog\"].isin(validation_test_df[\"GeneID\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AL6G25970</td>\n",
       "      <td>0</td>\n",
       "      <td>GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G15150</td>\n",
       "      <td>HOM04D000209</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2787</th>\n",
       "      <td>Carub.0005s1225</td>\n",
       "      <td>1</td>\n",
       "      <td>GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G46450</td>\n",
       "      <td>HOM04D003906</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2788</th>\n",
       "      <td>Carub.0005s2556</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58780</td>\n",
       "      <td>HOM04D000037</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2217 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "5           AL6G25970     0   \n",
       "...               ...   ...   \n",
       "2787  Carub.0005s1225     1   \n",
       "2788  Carub.0005s2556     1   \n",
       "2790  Carub.0006s3101     1   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "5     GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...     aly  AT5G15150   \n",
       "...                                                 ...     ...        ...   \n",
       "2787  GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...     cru  AT3G46450   \n",
       "2788  TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...     cru  AT3G58780   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "5     HOM04D000209  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "...            ...                                                ...   \n",
       "2787  HOM04D003906  [[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...   \n",
       "2788  HOM04D000037  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "5     [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "...                                                 ...  \n",
       "2787  [[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...  \n",
       "2788  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2217 rows x 8 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch time augmentation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def get_batch(data, phylo_aug_data, indices, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tCreates a batch of the input and one-hot encodes the sequences\n",
    "\t\"\"\"\n",
    "\tseqs = []\n",
    "\tlabels = []\n",
    "\tgene_ids = []\n",
    "\tfor ii in indices:\n",
    "\t\tgene_id = data.iloc[ii].GeneID\n",
    "\t\tsample = data.iloc[ii]\n",
    "\t\tif random.choices(population=[0, 1], weights=[1 - phylo_aug_rate, phylo_aug_rate])[0] == 1:\n",
    "\t\t\torthologs = phylo_aug_data[phylo_aug_data.Ortholog == gene_id]\n",
    "\t\t\tif not orthologs.empty:\n",
    "\t\t\t\tsample = phylo_aug_data[phylo_aug_data.Ortholog == gene_id].sample().iloc[0]\n",
    "\t\t\t# Randomly pick the forward or reverse complement strand\n",
    "\t\tif random.randint(0, 1) == 0:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.RC_one_hot_encoded)\n",
    "\t\telse:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.One_hot_encoded)\n",
    "\t\t\t\n",
    "\t\tlabels.append(sample.Label)\n",
    "\n",
    "\tX_batch = np.array(seqs)\n",
    "\tX_batch = X_batch.astype('float32')\n",
    "\n",
    "\tY_batch = np.array(labels)\n",
    "\tY_batch = Y_batch.astype('int64')\n",
    "\n",
    "\treturn X_batch, Y_batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_gen(data, phylo_aug_data, batch_size, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tGenerator function for loading input data in batches\n",
    "\t\"\"\"\n",
    "\tnum_samples = data.shape[0]\n",
    "\t# indices = list(range(num_samples))\n",
    "\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t# indices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\n",
    "\tii = 0\n",
    "\twhile True:\n",
    "\t\tyield get_batch(data, phylo_aug_data, indices[ii:ii + batch_size], perform_phylo_aug, phylo_aug_rate)\n",
    "\t\tii += batch_size\n",
    "\t\tif ii >= num_samples:\n",
    "\t\t\tii = 0\n",
    "\t\t\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t\t\t# indices = list(range(num_samples))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare validation input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_input_and_labels(df):\n",
    "\tohe_np = np.stack(df[\"One_hot_encoded\"])\n",
    "\trc_np = np.stack(df[\"RC_one_hot_encoded\"])\n",
    "\n",
    "\tx = np.append(ohe_np, rc_np, axis=0)\n",
    "\tx = x.astype('float32')\n",
    "\ty = np.append(df[\"Label\"], df[\"Label\"])\n",
    "\ty = y.astype('int64')\n",
    "\treturn x, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_validation, y_validation = get_input_and_labels(validation_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we do batch time augmentation we define the data generator here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "BATCH_SIZE = 200\n",
    "train_data_gen = data_gen(train_df,filtered_phylo_aug_df, BATCH_SIZE, True, 0.7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the shape of the input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_shape = train_df[\"One_hot_encoded\"].iloc[0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-05-21 15:56:19.690170: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "2025-05-21 15:56:19.690211: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "2025-05-21 15:56:19.690242: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (p-vibcompute-1.vib.local): /proc/driver/nvidia/version does not exist\n",
      "2025-05-21 15:56:19.690564: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    }
   ],
   "source": [
    "model = get_model(input_shape=input_shape, perform_evoaug=False,learning_rate=0.001)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# early stopping callback\n",
    "import tensorflow as tf\n",
    "\n",
    "early_stopping_callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\tpatience=40,\n",
    "\t\t\t\t\t\t\t\t\t\t\tverbose=1,\n",
    "\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\trestore_best_weights=True)\n",
    "# reduce learning rate callback\n",
    "reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tfactor=0.1,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tpatience=5,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmin_lr=1e-7,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tverbose=1)\n",
    "callbacks = [early_stopping_callback, reduce_lr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "10/10 [==============================] - 5s 305ms/step - loss: 0.6955 - acc: 0.4947 - auROC: 0.4725 - auPRC: 0.4758 - true_positives: 154.0000 - precision: 0.4695 - recall: 0.1638 - val_loss: 0.6933 - val_acc: 0.5039 - val_auROC: 0.5158 - val_auPRC: 0.5307 - val_true_positives: 249.0000 - val_precision: 0.5030 - val_recall: 0.9651 - lr: 0.0010\n",
      "Epoch 2/100\n",
      "10/10 [==============================] - 2s 246ms/step - loss: 0.6942 - acc: 0.4911 - auROC: 0.4936 - auPRC: 0.4847 - true_positives: 560.0000 - precision: 0.4882 - recall: 0.5957 - val_loss: 0.6930 - val_acc: 0.5233 - val_auROC: 0.5550 - val_auPRC: 0.5509 - val_true_positives: 32.0000 - val_precision: 0.6275 - val_recall: 0.1240 - lr: 0.0010\n",
      "Epoch 3/100\n",
      "10/10 [==============================] - 2s 239ms/step - loss: 0.6943 - acc: 0.5058 - auROC: 0.4866 - auPRC: 0.4937 - true_positives: 212.0000 - precision: 0.5012 - recall: 0.2255 - val_loss: 0.6934 - val_acc: 0.4981 - val_auROC: 0.5519 - val_auPRC: 0.5299 - val_true_positives: 0.0000e+00 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 4/100\n",
      "10/10 [==============================] - 2s 232ms/step - loss: 0.6925 - acc: 0.5105 - auROC: 0.5293 - auPRC: 0.5188 - true_positives: 216.0000 - precision: 0.5118 - recall: 0.2298 - val_loss: 0.6933 - val_acc: 0.5039 - val_auROC: 0.5272 - val_auPRC: 0.5212 - val_true_positives: 6.0000 - val_precision: 0.6667 - val_recall: 0.0233 - lr: 0.0010\n",
      "Epoch 5/100\n",
      "10/10 [==============================] - 2s 251ms/step - loss: 0.6935 - acc: 0.5205 - auROC: 0.5100 - auPRC: 0.5094 - true_positives: 381.0000 - precision: 0.5198 - recall: 0.4053 - val_loss: 0.6929 - val_acc: 0.5447 - val_auROC: 0.5435 - val_auPRC: 0.5277 - val_true_positives: 94.0000 - val_precision: 0.5732 - val_recall: 0.3643 - lr: 0.0010\n",
      "Epoch 6/100\n",
      "10/10 [==============================] - 2s 248ms/step - loss: 0.6934 - acc: 0.5068 - auROC: 0.5107 - auPRC: 0.4974 - true_positives: 257.0000 - precision: 0.5029 - recall: 0.2734 - val_loss: 0.6930 - val_acc: 0.4981 - val_auROC: 0.5810 - val_auPRC: 0.5613 - val_true_positives: 0.0000e+00 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 7/100\n",
      "10/10 [==============================] - 2s 240ms/step - loss: 0.6929 - acc: 0.5011 - auROC: 0.5157 - auPRC: 0.5036 - true_positives: 160.0000 - precision: 0.4878 - recall: 0.1702 - val_loss: 0.6923 - val_acc: 0.4981 - val_auROC: 0.5859 - val_auPRC: 0.5605 - val_true_positives: 0.0000e+00 - val_precision: 0.0000e+00 - val_recall: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 8/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.6917 - acc: 0.5295 - auROC: 0.5449 - auPRC: 0.5293 - true_positives: 195.0000 - precision: 0.5669 - recall: 0.2074 - val_loss: 0.6900 - val_acc: 0.5642 - val_auROC: 0.6205 - val_auPRC: 0.5869 - val_true_positives: 95.0000 - val_precision: 0.6090 - val_recall: 0.3682 - lr: 0.0010\n",
      "Epoch 9/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.6916 - acc: 0.5300 - auROC: 0.5394 - auPRC: 0.5184 - true_positives: 550.0000 - precision: 0.5223 - recall: 0.5851 - val_loss: 0.6910 - val_acc: 0.4981 - val_auROC: 0.6154 - val_auPRC: 0.5757 - val_true_positives: 3.0000 - val_precision: 0.5000 - val_recall: 0.0116 - lr: 0.0010\n",
      "Epoch 10/100\n",
      "10/10 [==============================] - 2s 238ms/step - loss: 0.6901 - acc: 0.5368 - auROC: 0.5500 - auPRC: 0.5424 - true_positives: 435.0000 - precision: 0.5370 - recall: 0.4628 - val_loss: 0.6826 - val_acc: 0.5759 - val_auROC: 0.6356 - val_auPRC: 0.6206 - val_true_positives: 111.0000 - val_precision: 0.6099 - val_recall: 0.4302 - lr: 0.0010\n",
      "Epoch 11/100\n",
      "10/10 [==============================] - 2s 248ms/step - loss: 0.6854 - acc: 0.5437 - auROC: 0.5746 - auPRC: 0.5480 - true_positives: 389.0000 - precision: 0.5518 - recall: 0.4138 - val_loss: 0.6953 - val_acc: 0.5097 - val_auROC: 0.6396 - val_auPRC: 0.6083 - val_true_positives: 256.0000 - val_precision: 0.5059 - val_recall: 0.9922 - lr: 0.0010\n",
      "Epoch 12/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.6820 - acc: 0.5674 - auROC: 0.5825 - auPRC: 0.5465 - true_positives: 560.0000 - precision: 0.5589 - recall: 0.5957 - val_loss: 0.6823 - val_acc: 0.5350 - val_auROC: 0.6463 - val_auPRC: 0.6294 - val_true_positives: 33.0000 - val_precision: 0.7021 - val_recall: 0.1279 - lr: 0.0010\n",
      "Epoch 13/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.6877 - acc: 0.5363 - auROC: 0.5537 - auPRC: 0.5388 - true_positives: 589.0000 - precision: 0.5264 - recall: 0.6266 - val_loss: 0.6797 - val_acc: 0.5798 - val_auROC: 0.6301 - val_auPRC: 0.6126 - val_true_positives: 209.0000 - val_precision: 0.5559 - val_recall: 0.8101 - lr: 0.0010\n",
      "Epoch 14/100\n",
      "10/10 [==============================] - 2s 252ms/step - loss: 0.6773 - acc: 0.5821 - auROC: 0.6107 - auPRC: 0.5804 - true_positives: 424.0000 - precision: 0.6040 - recall: 0.4511 - val_loss: 0.6796 - val_acc: 0.5739 - val_auROC: 0.6303 - val_auPRC: 0.6105 - val_true_positives: 230.0000 - val_precision: 0.5463 - val_recall: 0.8915 - lr: 0.0010\n",
      "Epoch 15/100\n",
      "10/10 [==============================] - 2s 228ms/step - loss: 0.6712 - acc: 0.5795 - auROC: 0.6204 - auPRC: 0.5866 - true_positives: 627.0000 - precision: 0.5633 - recall: 0.6670 - val_loss: 0.6646 - val_acc: 0.5992 - val_auROC: 0.6470 - val_auPRC: 0.6132 - val_true_positives: 203.0000 - val_precision: 0.5734 - val_recall: 0.7868 - lr: 0.0010\n",
      "Epoch 16/100\n",
      "10/10 [==============================] - 2s 235ms/step - loss: 0.6702 - acc: 0.5937 - auROC: 0.6138 - auPRC: 0.5732 - true_positives: 646.0000 - precision: 0.5747 - recall: 0.6872 - val_loss: 0.6634 - val_acc: 0.5934 - val_auROC: 0.6486 - val_auPRC: 0.6085 - val_true_positives: 176.0000 - val_precision: 0.5809 - val_recall: 0.6822 - lr: 0.0010\n",
      "Epoch 17/100\n",
      "10/10 [==============================] - 2s 238ms/step - loss: 0.6603 - acc: 0.6163 - auROC: 0.6485 - auPRC: 0.6079 - true_positives: 618.0000 - precision: 0.6029 - recall: 0.6574 - val_loss: 0.6611 - val_acc: 0.5973 - val_auROC: 0.6533 - val_auPRC: 0.6130 - val_true_positives: 198.0000 - val_precision: 0.5739 - val_recall: 0.7674 - lr: 0.0010\n",
      "Epoch 18/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.6603 - acc: 0.5984 - auROC: 0.6459 - auPRC: 0.6087 - true_positives: 654.0000 - precision: 0.5782 - recall: 0.6957 - val_loss: 0.6620 - val_acc: 0.5914 - val_auROC: 0.6557 - val_auPRC: 0.6304 - val_true_positives: 139.0000 - val_precision: 0.6043 - val_recall: 0.5388 - lr: 0.0010\n",
      "Epoch 19/100\n",
      "10/10 [==============================] - 2s 252ms/step - loss: 0.6589 - acc: 0.6126 - auROC: 0.6483 - auPRC: 0.6132 - true_positives: 598.0000 - precision: 0.6028 - recall: 0.6362 - val_loss: 0.6572 - val_acc: 0.6070 - val_auROC: 0.6602 - val_auPRC: 0.6240 - val_true_positives: 148.0000 - val_precision: 0.6167 - val_recall: 0.5736 - lr: 0.0010\n",
      "Epoch 20/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.6532 - acc: 0.6179 - auROC: 0.6568 - auPRC: 0.6238 - true_positives: 634.0000 - precision: 0.6015 - recall: 0.6745 - val_loss: 0.6653 - val_acc: 0.6089 - val_auROC: 0.6624 - val_auPRC: 0.6203 - val_true_positives: 123.0000 - val_precision: 0.6508 - val_recall: 0.4767 - lr: 0.0010\n",
      "Epoch 21/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.6495 - acc: 0.6216 - auROC: 0.6670 - auPRC: 0.6351 - true_positives: 636.0000 - precision: 0.6051 - recall: 0.6766 - val_loss: 0.6546 - val_acc: 0.6070 - val_auROC: 0.6639 - val_auPRC: 0.6187 - val_true_positives: 160.0000 - val_precision: 0.6061 - val_recall: 0.6202 - lr: 0.0010\n",
      "Epoch 22/100\n",
      "10/10 [==============================] - 2s 242ms/step - loss: 0.6353 - acc: 0.6484 - auROC: 0.6874 - auPRC: 0.6479 - true_positives: 638.0000 - precision: 0.6355 - recall: 0.6787 - val_loss: 0.6523 - val_acc: 0.6226 - val_auROC: 0.6722 - val_auPRC: 0.6299 - val_true_positives: 195.0000 - val_precision: 0.5982 - val_recall: 0.7558 - lr: 0.0010\n",
      "Epoch 23/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.6353 - acc: 0.6395 - auROC: 0.6895 - auPRC: 0.6600 - true_positives: 636.0000 - precision: 0.6254 - recall: 0.6766 - val_loss: 0.6477 - val_acc: 0.6187 - val_auROC: 0.6770 - val_auPRC: 0.6298 - val_true_positives: 186.0000 - val_precision: 0.6000 - val_recall: 0.7209 - lr: 0.0010\n",
      "Epoch 24/100\n",
      "10/10 [==============================] - 2s 228ms/step - loss: 0.6366 - acc: 0.6389 - auROC: 0.6892 - auPRC: 0.6637 - true_positives: 665.0000 - precision: 0.6180 - recall: 0.7074 - val_loss: 0.6597 - val_acc: 0.5973 - val_auROC: 0.6780 - val_auPRC: 0.6283 - val_true_positives: 217.0000 - val_precision: 0.5666 - val_recall: 0.8411 - lr: 0.0010\n",
      "Epoch 25/100\n",
      "10/10 [==============================] - 2s 236ms/step - loss: 0.6336 - acc: 0.6389 - auROC: 0.6918 - auPRC: 0.6572 - true_positives: 665.0000 - precision: 0.6180 - recall: 0.7074 - val_loss: 0.6518 - val_acc: 0.6401 - val_auROC: 0.6864 - val_auPRC: 0.6546 - val_true_positives: 137.0000 - val_precision: 0.6816 - val_recall: 0.5310 - lr: 0.0010\n",
      "Epoch 26/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.6325 - acc: 0.6374 - auROC: 0.6958 - auPRC: 0.6600 - true_positives: 623.0000 - precision: 0.6261 - recall: 0.6628 - val_loss: 0.6468 - val_acc: 0.6304 - val_auROC: 0.6841 - val_auPRC: 0.6496 - val_true_positives: 154.0000 - val_precision: 0.6417 - val_recall: 0.5969 - lr: 0.0010\n",
      "Epoch 27/100\n",
      "10/10 [==============================] - 2s 242ms/step - loss: 0.6108 - acc: 0.6658 - auROC: 0.7244 - auPRC: 0.6868 - true_positives: 683.0000 - precision: 0.6437 - recall: 0.7266 - val_loss: 0.6455 - val_acc: 0.6323 - val_auROC: 0.6862 - val_auPRC: 0.6517 - val_true_positives: 183.0000 - val_precision: 0.6162 - val_recall: 0.7093 - lr: 0.0010\n",
      "Epoch 28/100\n",
      "10/10 [==============================] - 2s 245ms/step - loss: 0.6180 - acc: 0.6563 - auROC: 0.7140 - auPRC: 0.6897 - true_positives: 622.0000 - precision: 0.6499 - recall: 0.6617 - val_loss: 0.6727 - val_acc: 0.6031 - val_auROC: 0.6709 - val_auPRC: 0.6305 - val_true_positives: 232.0000 - val_precision: 0.5659 - val_recall: 0.8992 - lr: 0.0010\n",
      "Epoch 29/100\n",
      "10/10 [==============================] - 2s 220ms/step - loss: 0.6146 - acc: 0.6716 - auROC: 0.7219 - auPRC: 0.6864 - true_positives: 701.0000 - precision: 0.6455 - recall: 0.7457 - val_loss: 0.6465 - val_acc: 0.6362 - val_auROC: 0.6817 - val_auPRC: 0.6465 - val_true_positives: 187.0000 - val_precision: 0.6172 - val_recall: 0.7248 - lr: 0.0010\n",
      "Epoch 30/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.6077 - acc: 0.6653 - auROC: 0.7311 - auPRC: 0.7137 - true_positives: 623.0000 - precision: 0.6614 - recall: 0.6628 - val_loss: 0.6435 - val_acc: 0.6284 - val_auROC: 0.6885 - val_auPRC: 0.6635 - val_true_positives: 164.0000 - val_precision: 0.6284 - val_recall: 0.6357 - lr: 0.0010\n",
      "Epoch 31/100\n",
      "10/10 [==============================] - 2s 245ms/step - loss: 0.5881 - acc: 0.6816 - auROC: 0.7575 - auPRC: 0.7281 - true_positives: 679.0000 - precision: 0.6637 - recall: 0.7223 - val_loss: 0.6556 - val_acc: 0.6245 - val_auROC: 0.6828 - val_auPRC: 0.6484 - val_true_positives: 205.0000 - val_precision: 0.5942 - val_recall: 0.7946 - lr: 0.0010\n",
      "Epoch 32/100\n",
      "10/10 [==============================] - 2s 230ms/step - loss: 0.5907 - acc: 0.6832 - auROC: 0.7489 - auPRC: 0.7253 - true_positives: 658.0000 - precision: 0.6728 - recall: 0.7000 - val_loss: 0.6477 - val_acc: 0.6304 - val_auROC: 0.6865 - val_auPRC: 0.6547 - val_true_positives: 164.0000 - val_precision: 0.6308 - val_recall: 0.6357 - lr: 0.0010\n",
      "Epoch 33/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.5857 - acc: 0.6821 - auROC: 0.7565 - auPRC: 0.7415 - true_positives: 655.0000 - precision: 0.6725 - recall: 0.6968 - val_loss: 0.6700 - val_acc: 0.6206 - val_auROC: 0.6673 - val_auPRC: 0.6160 - val_true_positives: 215.0000 - val_precision: 0.5858 - val_recall: 0.8333 - lr: 0.0010\n",
      "Epoch 34/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.5657 - acc: 0.7079 - auROC: 0.7790 - auPRC: 0.7526 - true_positives: 719.0000 - precision: 0.6828 - recall: 0.7649 - val_loss: 0.7020 - val_acc: 0.5953 - val_auROC: 0.6956 - val_auPRC: 0.6768 - val_true_positives: 92.0000 - val_precision: 0.6866 - val_recall: 0.3566 - lr: 0.0010\n",
      "Epoch 35/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5768 - acc: 0.7032 - auROC: 0.7651 - auPRC: 0.7375 - true_positives: 656.0000 - precision: 0.7009 - recall: 0.6979\n",
      "Epoch 35: ReduceLROnPlateau reducing learning rate to 0.00010000000474974513.\n",
      "10/10 [==============================] - 2s 224ms/step - loss: 0.5768 - acc: 0.7032 - auROC: 0.7651 - auPRC: 0.7375 - true_positives: 656.0000 - precision: 0.7009 - recall: 0.6979 - val_loss: 0.6561 - val_acc: 0.6362 - val_auROC: 0.6786 - val_auPRC: 0.6468 - val_true_positives: 154.0000 - val_precision: 0.6498 - val_recall: 0.5969 - lr: 0.0010\n",
      "Epoch 36/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.5550 - acc: 0.7200 - auROC: 0.7947 - auPRC: 0.7727 - true_positives: 647.0000 - precision: 0.7302 - recall: 0.6883 - val_loss: 0.6600 - val_acc: 0.6265 - val_auROC: 0.6692 - val_auPRC: 0.6333 - val_true_positives: 187.0000 - val_precision: 0.6071 - val_recall: 0.7248 - lr: 1.0000e-04\n",
      "Epoch 37/100\n",
      "10/10 [==============================] - 2s 226ms/step - loss: 0.5525 - acc: 0.7258 - auROC: 0.7971 - auPRC: 0.7681 - true_positives: 735.0000 - precision: 0.6993 - recall: 0.7819 - val_loss: 0.6585 - val_acc: 0.6206 - val_auROC: 0.6722 - val_auPRC: 0.6367 - val_true_positives: 177.0000 - val_precision: 0.6082 - val_recall: 0.6860 - lr: 1.0000e-04\n",
      "Epoch 38/100\n",
      "10/10 [==============================] - 2s 234ms/step - loss: 0.5549 - acc: 0.7226 - auROC: 0.7892 - auPRC: 0.7667 - true_positives: 676.0000 - precision: 0.7199 - recall: 0.7191 - val_loss: 0.6580 - val_acc: 0.6206 - val_auROC: 0.6758 - val_auPRC: 0.6420 - val_true_positives: 169.0000 - val_precision: 0.6145 - val_recall: 0.6550 - lr: 1.0000e-04\n",
      "Epoch 39/100\n",
      "10/10 [==============================] - 2s 229ms/step - loss: 0.5595 - acc: 0.7132 - auROC: 0.7858 - auPRC: 0.7595 - true_positives: 663.0000 - precision: 0.7121 - recall: 0.7053 - val_loss: 0.6637 - val_acc: 0.6265 - val_auROC: 0.6720 - val_auPRC: 0.6355 - val_true_positives: 189.0000 - val_precision: 0.6058 - val_recall: 0.7326 - lr: 1.0000e-04\n",
      "Epoch 40/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5448 - acc: 0.7289 - auROC: 0.8016 - auPRC: 0.7773 - true_positives: 732.0000 - precision: 0.7045 - recall: 0.7787\n",
      "Epoch 40: ReduceLROnPlateau reducing learning rate to 1.0000000474974514e-05.\n",
      "10/10 [==============================] - 2s 242ms/step - loss: 0.5448 - acc: 0.7289 - auROC: 0.8016 - auPRC: 0.7773 - true_positives: 732.0000 - precision: 0.7045 - recall: 0.7787 - val_loss: 0.6592 - val_acc: 0.6245 - val_auROC: 0.6773 - val_auPRC: 0.6451 - val_true_positives: 170.0000 - val_precision: 0.6182 - val_recall: 0.6589 - lr: 1.0000e-04\n",
      "Epoch 41/100\n",
      "10/10 [==============================] - 2s 228ms/step - loss: 0.5452 - acc: 0.7232 - auROC: 0.7994 - auPRC: 0.7785 - true_positives: 651.0000 - precision: 0.7331 - recall: 0.6926 - val_loss: 0.6592 - val_acc: 0.6245 - val_auROC: 0.6773 - val_auPRC: 0.6446 - val_true_positives: 170.0000 - val_precision: 0.6182 - val_recall: 0.6589 - lr: 1.0000e-05\n",
      "Epoch 42/100\n",
      "10/10 [==============================] - 2s 245ms/step - loss: 0.5537 - acc: 0.7258 - auROC: 0.7908 - auPRC: 0.7688 - true_positives: 676.0000 - precision: 0.7245 - recall: 0.7191 - val_loss: 0.6597 - val_acc: 0.6187 - val_auROC: 0.6766 - val_auPRC: 0.6420 - val_true_positives: 175.0000 - val_precision: 0.6076 - val_recall: 0.6783 - lr: 1.0000e-05\n",
      "Epoch 43/100\n",
      "10/10 [==============================] - 2s 246ms/step - loss: 0.5429 - acc: 0.7321 - auROC: 0.8013 - auPRC: 0.7786 - true_positives: 708.0000 - precision: 0.7188 - recall: 0.7532 - val_loss: 0.6606 - val_acc: 0.6206 - val_auROC: 0.6759 - val_auPRC: 0.6420 - val_true_positives: 178.0000 - val_precision: 0.6075 - val_recall: 0.6899 - lr: 1.0000e-05\n",
      "Epoch 44/100\n",
      "10/10 [==============================] - 2s 232ms/step - loss: 0.5424 - acc: 0.7337 - auROC: 0.8004 - auPRC: 0.7698 - true_positives: 705.0000 - precision: 0.7223 - recall: 0.7500 - val_loss: 0.6608 - val_acc: 0.6226 - val_auROC: 0.6755 - val_auPRC: 0.6414 - val_true_positives: 179.0000 - val_precision: 0.6088 - val_recall: 0.6938 - lr: 1.0000e-05\n",
      "Epoch 45/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5414 - acc: 0.7332 - auROC: 0.8020 - auPRC: 0.7756 - true_positives: 701.0000 - precision: 0.7234 - recall: 0.7457\n",
      "Epoch 45: ReduceLROnPlateau reducing learning rate to 1.0000000656873453e-06.\n",
      "10/10 [==============================] - 2s 245ms/step - loss: 0.5414 - acc: 0.7332 - auROC: 0.8020 - auPRC: 0.7756 - true_positives: 701.0000 - precision: 0.7234 - recall: 0.7457 - val_loss: 0.6611 - val_acc: 0.6245 - val_auROC: 0.6759 - val_auPRC: 0.6411 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-05\n",
      "Epoch 46/100\n",
      "10/10 [==============================] - 2s 242ms/step - loss: 0.5457 - acc: 0.7268 - auROC: 0.7992 - auPRC: 0.7758 - true_positives: 694.0000 - precision: 0.7177 - recall: 0.7383 - val_loss: 0.6612 - val_acc: 0.6245 - val_auROC: 0.6757 - val_auPRC: 0.6405 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-06\n",
      "Epoch 47/100\n",
      "10/10 [==============================] - 2s 241ms/step - loss: 0.5417 - acc: 0.7274 - auROC: 0.8019 - auPRC: 0.7812 - true_positives: 708.0000 - precision: 0.7123 - recall: 0.7532 - val_loss: 0.6611 - val_acc: 0.6245 - val_auROC: 0.6760 - val_auPRC: 0.6414 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-06\n",
      "Epoch 48/100\n",
      "10/10 [==============================] - 2s 248ms/step - loss: 0.5406 - acc: 0.7226 - auROC: 0.8038 - auPRC: 0.7784 - true_positives: 687.0000 - precision: 0.7149 - recall: 0.7309 - val_loss: 0.6611 - val_acc: 0.6245 - val_auROC: 0.6758 - val_auPRC: 0.6409 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-06\n",
      "Epoch 49/100\n",
      "10/10 [==============================] - 2s 244ms/step - loss: 0.5461 - acc: 0.7274 - auROC: 0.7963 - auPRC: 0.7715 - true_positives: 705.0000 - precision: 0.7136 - recall: 0.7500 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6413 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-06\n",
      "Epoch 50/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5422 - acc: 0.7247 - auROC: 0.8013 - auPRC: 0.7804 - true_positives: 701.0000 - precision: 0.7117 - recall: 0.7457\n",
      "Epoch 50: ReduceLROnPlateau reducing learning rate to 1.0000001111620805e-07.\n",
      "10/10 [==============================] - 2s 240ms/step - loss: 0.5422 - acc: 0.7247 - auROC: 0.8013 - auPRC: 0.7804 - true_positives: 701.0000 - precision: 0.7117 - recall: 0.7457 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6409 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-06\n",
      "Epoch 51/100\n",
      "10/10 [==============================] - 2s 235ms/step - loss: 0.5278 - acc: 0.7374 - auROC: 0.8168 - auPRC: 0.7997 - true_positives: 720.0000 - precision: 0.7207 - recall: 0.7660 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6409 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 52/100\n",
      "10/10 [==============================] - 2s 244ms/step - loss: 0.5433 - acc: 0.7305 - auROC: 0.8017 - auPRC: 0.7745 - true_positives: 685.0000 - precision: 0.7272 - recall: 0.7287 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6407 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 53/100\n",
      "10/10 [==============================] - 2s 239ms/step - loss: 0.5515 - acc: 0.7174 - auROC: 0.7928 - auPRC: 0.7675 - true_positives: 707.0000 - precision: 0.6993 - recall: 0.7521 - val_loss: 0.6611 - val_acc: 0.6245 - val_auROC: 0.6759 - val_auPRC: 0.6408 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 54/100\n",
      "10/10 [==============================] - 2s 232ms/step - loss: 0.5337 - acc: 0.7347 - auROC: 0.8116 - auPRC: 0.7994 - true_positives: 709.0000 - precision: 0.7220 - recall: 0.7543 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6408 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 55/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5427 - acc: 0.7284 - auROC: 0.8015 - auPRC: 0.7735 - true_positives: 708.0000 - precision: 0.7137 - recall: 0.7532\n",
      "Epoch 55: ReduceLROnPlateau reducing learning rate to 1e-07.\n",
      "10/10 [==============================] - 2s 244ms/step - loss: 0.5427 - acc: 0.7284 - auROC: 0.8015 - auPRC: 0.7735 - true_positives: 708.0000 - precision: 0.7137 - recall: 0.7532 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 56/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.5398 - acc: 0.7263 - auROC: 0.8031 - auPRC: 0.7793 - true_positives: 695.0000 - precision: 0.7165 - recall: 0.7394 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 57/100\n",
      "10/10 [==============================] - 2s 237ms/step - loss: 0.5547 - acc: 0.7284 - auROC: 0.7904 - auPRC: 0.7714 - true_positives: 708.0000 - precision: 0.7137 - recall: 0.7532 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 58/100\n",
      "10/10 [==============================] - 2s 247ms/step - loss: 0.5351 - acc: 0.7247 - auROC: 0.8089 - auPRC: 0.7932 - true_positives: 704.0000 - precision: 0.7104 - recall: 0.7489 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6759 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 59/100\n",
      "10/10 [==============================] - 2s 234ms/step - loss: 0.5498 - acc: 0.7274 - auROC: 0.7937 - auPRC: 0.7751 - true_positives: 689.0000 - precision: 0.7207 - recall: 0.7330 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 60/100\n",
      "10/10 [==============================] - 2s 246ms/step - loss: 0.5428 - acc: 0.7295 - auROC: 0.7992 - auPRC: 0.7680 - true_positives: 711.0000 - precision: 0.7139 - recall: 0.7564 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6410 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 61/100\n",
      "10/10 [==============================] - 2s 227ms/step - loss: 0.5450 - acc: 0.7179 - auROC: 0.7984 - auPRC: 0.7744 - true_positives: 684.0000 - precision: 0.7095 - recall: 0.7277 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6407 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 62/100\n",
      "10/10 [==============================] - 2s 240ms/step - loss: 0.5393 - acc: 0.7300 - auROC: 0.8051 - auPRC: 0.7891 - true_positives: 703.0000 - precision: 0.7181 - recall: 0.7479 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6405 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 63/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.5287 - acc: 0.7379 - auROC: 0.8143 - auPRC: 0.7940 - true_positives: 713.0000 - precision: 0.7246 - recall: 0.7585 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6405 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 64/100\n",
      "10/10 [==============================] - 2s 246ms/step - loss: 0.5398 - acc: 0.7305 - auROC: 0.8036 - auPRC: 0.7905 - true_positives: 715.0000 - precision: 0.7136 - recall: 0.7606 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6405 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 65/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.5430 - acc: 0.7242 - auROC: 0.7998 - auPRC: 0.7777 - true_positives: 684.0000 - precision: 0.7185 - recall: 0.7277 - val_loss: 0.6611 - val_acc: 0.6265 - val_auROC: 0.6757 - val_auPRC: 0.6404 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 66/100\n",
      "10/10 [==============================] - 2s 241ms/step - loss: 0.5345 - acc: 0.7300 - auROC: 0.8092 - auPRC: 0.7935 - true_positives: 697.0000 - precision: 0.7208 - recall: 0.7415 - val_loss: 0.6612 - val_acc: 0.6265 - val_auROC: 0.6757 - val_auPRC: 0.6404 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 67/100\n",
      "10/10 [==============================] - 2s 231ms/step - loss: 0.5389 - acc: 0.7337 - auROC: 0.8051 - auPRC: 0.7889 - true_positives: 704.0000 - precision: 0.7228 - recall: 0.7489 - val_loss: 0.6612 - val_acc: 0.6245 - val_auROC: 0.6757 - val_auPRC: 0.6404 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 68/100\n",
      "10/10 [==============================] - 2s 240ms/step - loss: 0.5468 - acc: 0.7205 - auROC: 0.7971 - auPRC: 0.7835 - true_positives: 697.0000 - precision: 0.7076 - recall: 0.7415 - val_loss: 0.6612 - val_acc: 0.6265 - val_auROC: 0.6757 - val_auPRC: 0.6404 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 69/100\n",
      "10/10 [==============================] - 2s 233ms/step - loss: 0.5400 - acc: 0.7195 - auROC: 0.8028 - auPRC: 0.7917 - true_positives: 686.0000 - precision: 0.7109 - recall: 0.7298 - val_loss: 0.6612 - val_acc: 0.6265 - val_auROC: 0.6757 - val_auPRC: 0.6404 - val_true_positives: 181.0000 - val_precision: 0.6115 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 70/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5456 - acc: 0.7253 - auROC: 0.7976 - auPRC: 0.7757 - true_positives: 689.0000 - precision: 0.7177 - recall: 0.7330Restoring model weights from the end of the best epoch: 30.\n",
      "10/10 [==============================] - 2s 230ms/step - loss: 0.5456 - acc: 0.7253 - auROC: 0.7976 - auPRC: 0.7757 - true_positives: 689.0000 - precision: 0.7177 - recall: 0.7330 - val_loss: 0.6612 - val_acc: 0.6245 - val_auROC: 0.6757 - val_auPRC: 0.6405 - val_true_positives: 181.0000 - val_precision: 0.6094 - val_recall: 0.7016 - lr: 1.0000e-07\n",
      "Epoch 70: early stopping\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "history = model.fit(train_data_gen,\n",
    "\t\t\t\t\tsteps_per_epoch=math.ceil(\n",
    "\t\t\t\t\t\ttrain_df.shape[0] / BATCH_SIZE),\n",
    "\t\t\t\t\tepochs=100,\n",
    "\t\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\t\tcallbacks=callbacks\n",
    "\t\t\t\t\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: Loss of 0.6434972286224365; accuracy of 0.6284046769142151; auroc of 0.6885219812393188; auprc of 0.6634527444839478; TP of 164.0\n"
     ]
    }
   ],
   "source": [
    "scores = model.evaluate(x_validation, y_validation, verbose=0)\n",
    "print(f'Score: Loss of {scores[0]}; accuracy of {scores[1]}; auroc of {scores[2]}; auprc of {scores[3]}; TP of {scores[4]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finetune phylo-aug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train_pure, y_train_pure = get_input_and_labels(train_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n",
      "38/38 [==============================] - 4s 63ms/step - loss: 0.5985 - acc: 0.6824 - auROC: 0.7413 - auPRC: 0.7063 - true_positives_1: 1313.0000 - precision_1: 0.6723 - recall_1: 0.6984 - val_loss: 0.6441 - val_acc: 0.6401 - val_auROC: 0.6875 - val_auPRC: 0.6582 - val_true_positives_1: 185.0000 - val_precision_1: 0.6229 - val_recall_1: 0.7171\n",
      "Epoch 2/10\n",
      "38/38 [==============================] - 2s 50ms/step - loss: 0.5886 - acc: 0.6882 - auROC: 0.7541 - auPRC: 0.7171 - true_positives_1: 1344.0000 - precision_1: 0.6744 - recall_1: 0.7149 - val_loss: 0.6474 - val_acc: 0.6459 - val_auROC: 0.6875 - val_auPRC: 0.6544 - val_true_positives_1: 196.0000 - val_precision_1: 0.6203 - val_recall_1: 0.7597\n",
      "Epoch 3/10\n",
      "38/38 [==============================] - 2s 49ms/step - loss: 0.5826 - acc: 0.6908 - auROC: 0.7604 - auPRC: 0.7246 - true_positives_1: 1355.0000 - precision_1: 0.6758 - recall_1: 0.7207 - val_loss: 0.6449 - val_acc: 0.6381 - val_auROC: 0.6902 - val_auPRC: 0.6583 - val_true_positives_1: 186.0000 - val_precision_1: 0.6200 - val_recall_1: 0.7209\n",
      "Epoch 4/10\n",
      "38/38 [==============================] - 2s 51ms/step - loss: 0.5834 - acc: 0.6945 - auROC: 0.7582 - auPRC: 0.7261 - true_positives_1: 1379.0000 - precision_1: 0.6763 - recall_1: 0.7335 - val_loss: 0.6505 - val_acc: 0.6576 - val_auROC: 0.6867 - val_auPRC: 0.6497 - val_true_positives_1: 203.0000 - val_precision_1: 0.6265 - val_recall_1: 0.7868\n",
      "Epoch 5/10\n",
      "38/38 [==============================] - 2s 50ms/step - loss: 0.5747 - acc: 0.6989 - auROC: 0.7675 - auPRC: 0.7407 - true_positives_1: 1352.0000 - precision_1: 0.6870 - recall_1: 0.7191 - val_loss: 0.6575 - val_acc: 0.6420 - val_auROC: 0.6829 - val_auPRC: 0.6438 - val_true_positives_1: 209.0000 - val_precision_1: 0.6076 - val_recall_1: 0.8101\n",
      "Epoch 6/10\n",
      "38/38 [==============================] - 2s 52ms/step - loss: 0.5746 - acc: 0.6989 - auROC: 0.7673 - auPRC: 0.7383 - true_positives_1: 1391.0000 - precision_1: 0.6799 - recall_1: 0.7399 - val_loss: 0.6462 - val_acc: 0.6381 - val_auROC: 0.6916 - val_auPRC: 0.6595 - val_true_positives_1: 172.0000 - val_precision_1: 0.6324 - val_recall_1: 0.6667\n",
      "Epoch 7/10\n",
      "38/38 [==============================] - 2s 50ms/step - loss: 0.5636 - acc: 0.7158 - auROC: 0.7791 - auPRC: 0.7529 - true_positives_1: 1383.0000 - precision_1: 0.7035 - recall_1: 0.7356 - val_loss: 0.6520 - val_acc: 0.6420 - val_auROC: 0.6868 - val_auPRC: 0.6508 - val_true_positives_1: 196.0000 - val_precision_1: 0.6164 - val_recall_1: 0.7597\n",
      "Epoch 8/10\n",
      "38/38 [==============================] - 2s 52ms/step - loss: 0.5653 - acc: 0.7111 - auROC: 0.7763 - auPRC: 0.7443 - true_positives_1: 1412.0000 - precision_1: 0.6915 - recall_1: 0.7511 - val_loss: 0.6530 - val_acc: 0.6459 - val_auROC: 0.6865 - val_auPRC: 0.6478 - val_true_positives_1: 196.0000 - val_precision_1: 0.6203 - val_recall_1: 0.7597\n",
      "Epoch 9/10\n",
      "38/38 [==============================] - 2s 50ms/step - loss: 0.5641 - acc: 0.7187 - auROC: 0.7803 - auPRC: 0.7476 - true_positives_1: 1389.0000 - precision_1: 0.7062 - recall_1: 0.7388 - val_loss: 0.6493 - val_acc: 0.6323 - val_auROC: 0.6897 - val_auPRC: 0.6551 - val_true_positives_1: 173.0000 - val_precision_1: 0.6245 - val_recall_1: 0.6705\n",
      "Epoch 10/10\n",
      "38/38 [==============================] - 2s 52ms/step - loss: 0.5524 - acc: 0.7211 - auROC: 0.7903 - auPRC: 0.7657 - true_positives_1: 1412.0000 - precision_1: 0.7046 - recall_1: 0.7511 - val_loss: 0.6563 - val_acc: 0.6420 - val_auROC: 0.6844 - val_auPRC: 0.6437 - val_true_positives_1: 199.0000 - val_precision_1: 0.6142 - val_recall_1: 0.7713\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.callbacks.History at 0x7fe7580765e0>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tensorflow import keras\n",
    "from src.diff_expression_model import compile_model\n",
    "\n",
    "# set up callbacks\n",
    "model = compile_model(model, learning_rate=0.0001)\n",
    "\n",
    "# train model\n",
    "model.fit(x_train_pure, y_train_pure,\n",
    "\t\t\t\tepochs=10,\n",
    "\t\t\t\tbatch_size=100,\n",
    "\t\t\t\tshuffle=True,\n",
    "\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\tcallbacks=[early_stopping_callback])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: Loss of 0.6562544703483582; accuracy of 0.6420233249664307; auroc of 0.6844037771224976; auprc of 0.6436824202537537; TP of 199.0\n"
     ]
    }
   ],
   "source": [
    "scores = model.evaluate(x_validation, y_validation, verbose=0)\n",
    "print(f'Score: Loss of {scores[0]}; accuracy of {scores[1]}; auroc of {scores[2]}; auprc of {scores[3]}; TP of {scores[4]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We perform post-hoc conjoing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "siamese_model = get_siamese_model(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare the test set data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test = np.stack(test_df[\"One_hot_encoded\"])\n",
    "x_test_rc = np.stack(test_df[\"RC_one_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_test = test_df[\"Label\"].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8/8 [==============================] - 0s 8ms/step\n"
     ]
    }
   ],
   "source": [
    "predictions_categories, predictions = post_hoc_conjoining(siamese_model, x_test, x_test_rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7494802494802495"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_auroc(y_test, predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.70      0.56      0.62       111\n",
      "           1       0.68      0.80      0.73       130\n",
      "\n",
      "    accuracy                           0.69       241\n",
      "   macro avg       0.69      0.68      0.68       241\n",
      "weighted avg       0.69      0.69      0.68       241\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(y_test, predictions_categories))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
