{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/DeepDifE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-05 15:35:18.965727: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-08-05 15:35:19.024550: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-08-05 15:35:19.026087: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-08-05 15:35:20.043064: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%cd ..\n",
    "\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from evoaug_tf import evoaug, augment\n",
    "from src.diff_expression_model import get_model, get_siamese_model, post_hoc_conjoining, get_auroc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now we start from this data pickle as I'm not aware how Helder did DE analysis and generated the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppath = \"data/dataset_solid_chrome.pkl\"\n",
    "with open(ppath, 'rb') as f:\n",
    "    data = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>seqs</th>\n",
       "      <th>ohs</th>\n",
       "      <th>rcohs</th>\n",
       "      <th>ohsDuo</th>\n",
       "      <th>in_original_balanced</th>\n",
       "      <th>set</th>\n",
       "      <th>npshap-single</th>\n",
       "      <th>npshap-posthoc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AT4G27120</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[-0.00029412], [0.00011191], [0.00071743], ...</td>\n",
       "      <td>[[[[-0.00024759], [0.00013361], [0.00015087], ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT4G19600</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[0.00049611], [0.00069235], [0.00377408], [...</td>\n",
       "      <td>[[[[0.00051932], [0.00032187], [0.00027891], [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT3G60880</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[-0.00030243], [-1.08283739e-05], [0.000359...</td>\n",
       "      <td>[[[[-0.00031496], [-8.34961975e-05], [0.000132...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G06960</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[0.00254294], [0.00639127], [0.00152307], [...</td>\n",
       "      <td>[[[[0.00208253], [0.0018626], [0.00023493], [0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G14890</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[-5.23597009e-05], [0.00045373], [0.0003575...</td>\n",
       "      <td>[[[[0.00051831], [0.00050734], [4.30682958e-05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G64230</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[[[[-1.85048248e-05], [5.31432997e-05], [0.000...</td>\n",
       "      <td>[[[[0.00034249], [0.00039158], [1.82568183e-05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G64780</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[9.53665585e-05], [0.00032812], [0.00046541...</td>\n",
       "      <td>[[[[-5.87131144e-05], [6.68894427e-05], [6.172...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT4G30470</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[[[[0.00061825], [6.60565856e-05], [5.89712399...</td>\n",
       "      <td>[[[[0.00044855], [6.3632986e-05], [0.00018469]...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT3G51895</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[-0.00021527], [0.00037615], [0.0002829], [...</td>\n",
       "      <td>[[[[-0.00010363], [0.00015631], [0.0001], [0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT2G35585</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[0.00069553], [0.00023614], [0.00080674], [...</td>\n",
       "      <td>[[[[0.00047114], [9.892659e-05], [0.00010585],...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Category    GeneFamily  \\\n",
       "geneID                             \n",
       "AT4G27120        0  HOM04D000881   \n",
       "AT4G19600        0  HOM04D000740   \n",
       "AT3G60880        0  HOM04D003119   \n",
       "AT5G06960        0  HOM04D000319   \n",
       "AT1G14890        0  HOM04D000273   \n",
       "...            ...           ...   \n",
       "AT5G64230        1  HOM04D003278   \n",
       "AT5G64780        1  HOM04D002552   \n",
       "AT4G30470        1  HOM04D000082   \n",
       "AT3G51895        1  HOM04D000270   \n",
       "AT2G35585        1  HOM04D004118   \n",
       "\n",
       "                                                        seqs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "AT4G19600  GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "AT3G60880  AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "AT5G06960  CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "AT1G14890  TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                      ...   \n",
       "AT5G64230  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "AT5G64780  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "AT4G30470  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "AT3G51895  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "AT2G35585  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                                         ohs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT4G19600  [[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...   \n",
       "AT3G60880  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "AT5G06960  [[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "AT1G14890  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT5G64780  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "AT4G30470  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "AT3G51895  [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT2G35585  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "\n",
       "                                                       rcohs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT4G19600  [[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT3G60880  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT5G06960  [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT1G14890  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "AT5G64780  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT4G30470  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT3G51895  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "AT2G35585  [[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...   \n",
       "\n",
       "                                                      ohsDuo  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1...   \n",
       "AT4G19600  [[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1...   \n",
       "AT3G60880  [[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0...   \n",
       "AT5G06960  [[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0...   \n",
       "AT1G14890  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1...   \n",
       "AT5G64780  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...   \n",
       "AT4G30470  [[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0...   \n",
       "AT3G51895  [[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0...   \n",
       "AT2G35585  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...   \n",
       "\n",
       "           in_original_balanced    set  \\\n",
       "geneID                                   \n",
       "AT4G27120                  True  train   \n",
       "AT4G19600                  True  valid   \n",
       "AT3G60880                  True  valid   \n",
       "AT5G06960                  True  train   \n",
       "AT1G14890                  True  valid   \n",
       "...                         ...    ...   \n",
       "AT5G64230                  True   test   \n",
       "AT5G64780                  True  train   \n",
       "AT4G30470                  True   test   \n",
       "AT3G51895                  True  train   \n",
       "AT2G35585                  True  train   \n",
       "\n",
       "                                               npshap-single  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[[[-0.00029412], [0.00011191], [0.00071743], ...   \n",
       "AT4G19600  [[[[0.00049611], [0.00069235], [0.00377408], [...   \n",
       "AT3G60880  [[[[-0.00030243], [-1.08283739e-05], [0.000359...   \n",
       "AT5G06960  [[[[0.00254294], [0.00639127], [0.00152307], [...   \n",
       "AT1G14890  [[[[-5.23597009e-05], [0.00045373], [0.0003575...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[[[-1.85048248e-05], [5.31432997e-05], [0.000...   \n",
       "AT5G64780  [[[[9.53665585e-05], [0.00032812], [0.00046541...   \n",
       "AT4G30470  [[[[0.00061825], [6.60565856e-05], [5.89712399...   \n",
       "AT3G51895  [[[[-0.00021527], [0.00037615], [0.0002829], [...   \n",
       "AT2G35585  [[[[0.00069553], [0.00023614], [0.00080674], [...   \n",
       "\n",
       "                                              npshap-posthoc  \n",
       "geneID                                                        \n",
       "AT4G27120  [[[[-0.00024759], [0.00013361], [0.00015087], ...  \n",
       "AT4G19600  [[[[0.00051932], [0.00032187], [0.00027891], [...  \n",
       "AT3G60880  [[[[-0.00031496], [-8.34961975e-05], [0.000132...  \n",
       "AT5G06960  [[[[0.00208253], [0.0018626], [0.00023493], [0...  \n",
       "AT1G14890  [[[[0.00051831], [0.00050734], [4.30682958e-05...  \n",
       "...                                                      ...  \n",
       "AT5G64230  [[[[0.00034249], [0.00039158], [1.82568183e-05...  \n",
       "AT5G64780  [[[[-5.87131144e-05], [6.68894427e-05], [6.172...  \n",
       "AT4G30470  [[[[0.00044855], [6.3632986e-05], [0.00018469]...  \n",
       "AT3G51895  [[[[-0.00010363], [0.00015631], [0.0001], [0.0...  \n",
       "AT2G35585  [[[[0.00047114], [9.892659e-05], [0.00010585],...  \n",
       "\n",
       "[2398 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "train    1900\n",
       "valid     257\n",
       "test      241\n",
       "Name: set, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"set\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To show how to subdivide the dataset into train-test split we only take the following columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = data.reset_index()\n",
    "dataset = dataset[[\"geneID\", \"Category\", \"GeneFamily\", \"seqs\"]]\n",
    "dataset.rename(columns={\"geneID\":\"GeneID\", \"Category\":\"Label\", \"seqs\": \"Sequence\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID Label    GeneFamily  \\\n",
       "0     AT4G27120     0  HOM04D000881   \n",
       "1     AT4G19600     0  HOM04D000740   \n",
       "2     AT3G60880     0  HOM04D003119   \n",
       "3     AT5G06960     0  HOM04D000319   \n",
       "4     AT1G14890     0  HOM04D000273   \n",
       "...         ...   ...           ...   \n",
       "2393  AT5G64230     1  HOM04D003278   \n",
       "2394  AT5G64780     1  HOM04D002552   \n",
       "2395  AT4G30470     1  HOM04D000082   \n",
       "2396  AT3G51895     1  HOM04D000270   \n",
       "2397  AT2G35585     1  HOM04D004118   \n",
       "\n",
       "                                               Sequence  \n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...  \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...  \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...  \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...  \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...  \n",
       "...                                                 ...  \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...  \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...  \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...  \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...  \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...  \n",
       "\n",
       "[2398 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylo augmentation "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the fasta files for each species of which we want to use orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_augmentation_base_path = \"data/phylo_aug/fasta/\"\n",
    "species_fasta_files = {\n",
    "\t\t\t\"aly\":\"aly.tss.fasta\",\n",
    "\t\t\t\"chi\":\"chi.tss.fasta\",\n",
    "\t\t\t\"cru\":\"cru.tss.fasta\"\n",
    "}\n",
    "\n",
    "species_to_include = [\"aly\",\"chi\",\"cru\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsing the fasta files. In this case the label of the original sequence is incorporated in the ids in the fasta file\n",
    "e.g:\n",
    "```\n",
    ">Aa31LG6G5540_1\n",
    "TTGTATAAATATATTAAATT\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: data/phylo_aug/fasta/aly.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/chi.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/cru.tss.fasta done!\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequence_list = []\n",
    "gene_id_list = []\n",
    "label_list = []\n",
    "species_list = []\n",
    "for species in species_to_include:\n",
    "\tspecies_fasta = species_fasta_files[species]\n",
    "\tfasta_path = f\"{phylo_augmentation_base_path}{species_fasta}\"\n",
    "\tfasta_sequences = SeqIO.parse(open(fasta_path),'fasta')\n",
    "\tprint(f\"File: {fasta_path} done!\")\n",
    "\tfor fasta in fasta_sequences:\n",
    "\t\tname, sequence = fasta.id, str(fasta.seq)\n",
    "\t\t\n",
    "\t\t# The label is stored inside the id name (e.g. >ATERI-4G42060_1), split these\n",
    "\t\tgene_id = name.split(\"_\")[0]\n",
    "\t\tlabel = name.split(\"_\")[1]\n",
    "\t\tspecies = species_fasta.split(\".\")[0]\n",
    "\n",
    "\t\tgene_id_list.append(gene_id)\n",
    "\t\tsequence_list.append(sequence)\n",
    "\t\tlabel_list.append(label)\n",
    "\t\tspecies_list.append(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.DataFrame({\"GeneID\": gene_id_list, \"Label\": label_list, \"Sequence\": sequence_list, \"Species\": species_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  \n",
       "...                                                 ...     ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  \n",
       "\n",
       "[2794 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get gene family of the original ortholog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because during the batch time augmentation a original sample will be swapped by one of its orthologs, we need to know from which original sample each ortholog originates from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir_ortholog = \"data/phylo_aug/orthologs/\"\n",
    "ortholog_files = {\n",
    "\t\t\t\"ath-aly\":\"ath_aly.txt\",\n",
    "\t\t\t\"ath-chi\":\"ath_chi.txt\",\n",
    "\t\t\t\"ath-cru\":\"ath_cru.txt\"\n",
    "\t\t\t}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a lookup table to include the orthologs in the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "lookup_table = {}\n",
    "for orthologs in ortholog_files.values():\n",
    "\twith open(f\"{base_dir_ortholog}{orthologs}\", 'r') as file:\n",
    "\t\tfor line in file:\n",
    "\t\t\t# Split each line by tab\n",
    "\t\t\tcolumns = line.strip().split('\\t')\n",
    "\t\t\tif len(columns) >= 2:  # Ensure there are at least two columns\n",
    "\t\t\t\tkey = columns[1]\n",
    "\t\t\t\tvalue = columns[0]\n",
    "\t\t\t\tlookup_table[key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df[\"Ortholog\"] = phylo_aug_df[\"GeneID\"].map(lookup_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260  \n",
       "...                                                 ...     ...        ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970  \n",
       "\n",
       "[2794 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.merge(phylo_aug_df, dataset[[\"GeneID\", \"GeneFamily\"]], left_on=\"Ortholog\", right_on=\"GeneID\", how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2213</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2214</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2216</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2217</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2218 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2213  Carub.0004s0553     0   \n",
       "2214  Carub.0006s3101     1   \n",
       "2215  Carub.0005s2520     0   \n",
       "2216  Carub.0005s2528     0   \n",
       "2217  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2213  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2214  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2215  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2216  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2217  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily  \n",
       "0     HOM04D003099  \n",
       "1     HOM04D003060  \n",
       "2     HOM04D000135  \n",
       "3     HOM04D001302  \n",
       "4     HOM04D006130  \n",
       "...            ...  \n",
       "2213  HOM04D004133  \n",
       "2214  HOM04D001924  \n",
       "2215  HOM04D006357  \n",
       "2216  HOM04D007279  \n",
       "2217  HOM04D001540  \n",
       "\n",
       "[2218 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df.drop(\"GeneID_y\", axis=1, inplace=True)\n",
    "phylo_aug_df.rename(columns={\"GeneID_x\": \"GeneID\"}, inplace=True)\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One-hot-encode & reverse-complement both dataset as augmented dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import one_hot_encode_series, reverse_complement_series, reverse_complement_sequence\n",
    "dataset[\"One_hot_encoded\"] = one_hot_encode_series(dataset[\"Sequence\"])\n",
    "phylo_aug_df[\"One_hot_encoded\"] = one_hot_encode_series(phylo_aug_df[\"Sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[\"RC_one_hot_encoded\"] = reverse_complement_series(dataset[\"One_hot_encoded\"])\n",
    "phylo_aug_df[\"RC_one_hot_encoded\"] = reverse_complement_series(phylo_aug_df[\"One_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID Label    GeneFamily  \\\n",
       "0     AT4G27120     0  HOM04D000881   \n",
       "1     AT4G19600     0  HOM04D000740   \n",
       "2     AT3G60880     0  HOM04D003119   \n",
       "3     AT5G06960     0  HOM04D000319   \n",
       "4     AT1G14890     0  HOM04D000273   \n",
       "...         ...   ...           ...   \n",
       "2393  AT5G64230     1  HOM04D003278   \n",
       "2394  AT5G64780     1  HOM04D002552   \n",
       "2395  AT4G30470     1  HOM04D000082   \n",
       "2396  AT3G51895     1  HOM04D000270   \n",
       "2397  AT2G35585     1  HOM04D004118   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                 ...   \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                        One_hot_encoded  \\\n",
       "0     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "1     [[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...   \n",
       "2     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "3     [[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "4     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "...                                                 ...   \n",
       "2393  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "2394  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "2395  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2396  [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "2397  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "1     [[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2     [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "3     [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...  \n",
       "4     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...  \n",
       "...                                                 ...  \n",
       "2393  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...  \n",
       "2394  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2395  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "2396  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...  \n",
       "2397  [[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "\n",
       "[2398 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "                                        One_hot_encoded  \\\n",
       "0     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...                                                 ...   \n",
       "2789  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2790  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2791  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2792  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2789  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2791  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2794 rows x 7 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2213</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2214</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2216</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2217</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2218 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2213  Carub.0004s0553     0   \n",
       "2214  Carub.0006s3101     1   \n",
       "2215  Carub.0005s2520     0   \n",
       "2216  Carub.0005s2528     0   \n",
       "2217  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2213  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2214  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2215  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2216  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2217  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     HOM04D006130  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...            ...                                                ...   \n",
       "2213  HOM04D004133  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2214  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2215  HOM04D006357  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2216  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2217  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2213  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2214  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2215  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2216  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2217  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2218 rows x 8 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df = phylo_aug_df[phylo_aug_df[\"GeneFamily\"].notna()]\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train-test split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import grouped_shuffle_split\n",
    "train_df, validation_test_df = grouped_shuffle_split(dataset, dataset[\"GeneFamily\"], 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_df, test_df  = grouped_shuffle_split(validation_test_df, validation_test_df[\"GeneFamily\"], 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of training set: 1900\n",
      "Length of validation set: 257\n",
      "Length of test set: 241\n"
     ]
    }
   ],
   "source": [
    "print(f\"Length of training set: {train_df.shape[0]}\")\n",
    "print(f\"Length of validation set: {validation_df.shape[0]}\")\n",
    "print(f\"Length of test set: {test_df.shape[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Append train and augmentation dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter out the orthologs from samples in test or validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_phylo_aug_df = phylo_aug_df[~phylo_aug_df[\"Ortholog\"].isin(validation_test_df[\"GeneID\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AL6G25970</td>\n",
       "      <td>0</td>\n",
       "      <td>GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G15150</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2787</th>\n",
       "      <td>Carub.0005s1225</td>\n",
       "      <td>1</td>\n",
       "      <td>GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G46450</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2788</th>\n",
       "      <td>Carub.0005s2556</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58780</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2217 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "5           AL6G25970     0   \n",
       "...               ...   ...   \n",
       "2787  Carub.0005s1225     1   \n",
       "2788  Carub.0005s2556     1   \n",
       "2790  Carub.0006s3101     1   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "5     GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...     aly  AT5G15150   \n",
       "...                                                 ...     ...        ...   \n",
       "2787  GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...     cru  AT3G46450   \n",
       "2788  TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...     cru  AT3G58780   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "                                        One_hot_encoded  \\\n",
       "0     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "5     [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "...                                                 ...   \n",
       "2787  [[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...   \n",
       "2788  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "2790  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2792  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "5     [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "...                                                 ...  \n",
       "2787  [[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...  \n",
       "2788  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2217 rows x 7 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch time augmentation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def get_batch(data, phylo_aug_data, indices, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tCreates a batch of the input and one-hot encodes the sequences\n",
    "\t\"\"\"\n",
    "\tseqs = []\n",
    "\tlabels = []\n",
    "\tgene_ids = []\n",
    "\tfor ii in indices:\n",
    "\t\tgene_id = data.iloc[ii].GeneID\n",
    "\t\tsample = data.iloc[ii]\n",
    "\t\tif random.choices(population=[0, 1], weights=[1 - phylo_aug_rate, phylo_aug_rate])[0] == 1:\n",
    "\t\t\torthologs = phylo_aug_data[phylo_aug_data.Ortholog == gene_id]\n",
    "\t\t\tif not orthologs.empty:\n",
    "\t\t\t\tsample = phylo_aug_data[phylo_aug_data.Ortholog == gene_id].sample().iloc[0]\n",
    "\t\t\t# Randomly pick the forward or reverse complement strand\n",
    "\t\tif random.randint(0, 1) == 0:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.RC_one_hot_encoded)\n",
    "\t\telse:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.One_hot_encoded)\n",
    "\t\t\t\n",
    "\t\tlabels.append(sample.Label)\n",
    "\n",
    "\tX_batch = np.array(seqs)\n",
    "\tX_batch = X_batch.astype('float32')\n",
    "\n",
    "\tY_batch = np.array(labels)\n",
    "\tY_batch = Y_batch.astype('int64')\n",
    "\n",
    "\treturn X_batch, Y_batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_gen(data, phylo_aug_data, batch_size, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tGenerator function for loading input data in batches\n",
    "\t\"\"\"\n",
    "\tnum_samples = data.shape[0]\n",
    "\t# indices = list(range(num_samples))\n",
    "\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t# indices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\n",
    "\tii = 0\n",
    "\twhile True:\n",
    "\t\tyield get_batch(data, phylo_aug_data, indices[ii:ii + batch_size], perform_phylo_aug, phylo_aug_rate)\n",
    "\t\tii += batch_size\n",
    "\t\tif ii >= num_samples:\n",
    "\t\t\tii = 0\n",
    "\t\t\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t\t\t# indices = list(range(num_samples))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare validation input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_input_and_labels(df):\n",
    "\tohe_np = np.stack(df[\"One_hot_encoded\"])\n",
    "\trc_np = np.stack(df[\"RC_one_hot_encoded\"])\n",
    "\n",
    "\tx = np.append(ohe_np, rc_np, axis=0)\n",
    "\tx = x.astype('float32')\n",
    "\ty = np.append(df[\"Label\"], df[\"Label\"])\n",
    "\ty = y.astype('int64')\n",
    "\treturn x, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_validation, y_validation = get_input_and_labels(validation_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we do batch time augmentation we define the data generator here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "BATCH_SIZE = 200\n",
    "train_data_gen = data_gen(train_df,filtered_phylo_aug_df, BATCH_SIZE, True, 0.7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the shape of the input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_shape = train_df[\"One_hot_encoded\"].iloc[0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-05 15:51:24.365364: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:995] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero. See more at https://github.com/torvalds/linux/blob/v6.0/Documentation/ABI/testing/sysfs-bus-pci#L344-L355\n",
      "2024-08-05 15:51:24.368968: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1960] Cannot dlopen some GPU libraries. Please make sure the missing libraries mentioned above are installed properly if you would like to use GPU. Follow the guide at https://www.tensorflow.org/install/gpu for how to download and setup the required libraries for your platform.\n",
      "Skipping registering GPU devices...\n"
     ]
    }
   ],
   "source": [
    "model = get_model(input_shape=input_shape, perform_evoaug=False,learning_rate=0.001)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# early stopping callback\n",
    "import tensorflow as tf\n",
    "\n",
    "early_stopping_callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\tpatience=40,\n",
    "\t\t\t\t\t\t\t\t\t\t\tverbose=1,\n",
    "\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\trestore_best_weights=True)\n",
    "# reduce learning rate callback\n",
    "reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tfactor=0.1,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tpatience=5,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmin_lr=1e-7,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tverbose=1)\n",
    "callbacks = [early_stopping_callback, reduce_lr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "10/10 [==============================] - 4s 270ms/step - loss: 0.6948 - acc: 0.4832 - auROC: 0.4836 - auPRC: 0.4845 - true_positives: 574.0000 - val_loss: 0.6937 - val_acc: 0.4981 - val_auROC: 0.5483 - val_auPRC: 0.5562 - val_true_positives: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 2/100\n",
      "10/10 [==============================] - 2s 193ms/step - loss: 0.6943 - acc: 0.5163 - auROC: 0.4913 - auPRC: 0.4982 - true_positives: 180.0000 - val_loss: 0.6932 - val_acc: 0.5136 - val_auROC: 0.5558 - val_auPRC: 0.5362 - val_true_positives: 248.0000 - lr: 0.0010\n",
      "Epoch 3/100\n",
      "10/10 [==============================] - 2s 189ms/step - loss: 0.6939 - acc: 0.5042 - auROC: 0.4987 - auPRC: 0.4997 - true_positives: 289.0000 - val_loss: 0.6930 - val_acc: 0.4961 - val_auROC: 0.5664 - val_auPRC: 0.5521 - val_true_positives: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 4/100\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6931 - acc: 0.5147 - auROC: 0.5184 - auPRC: 0.5134 - true_positives: 172.0000 - val_loss: 0.6924 - val_acc: 0.5700 - val_auROC: 0.5725 - val_auPRC: 0.5492 - val_true_positives: 195.0000 - lr: 0.0010\n",
      "Epoch 5/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6931 - acc: 0.5158 - auROC: 0.5183 - auPRC: 0.5032 - true_positives: 653.0000 - val_loss: 0.6916 - val_acc: 0.5272 - val_auROC: 0.6074 - val_auPRC: 0.5652 - val_true_positives: 64.0000 - lr: 0.0010\n",
      "Epoch 6/100\n",
      "10/10 [==============================] - 2s 196ms/step - loss: 0.6912 - acc: 0.5221 - auROC: 0.5448 - auPRC: 0.5297 - true_positives: 195.0000 - val_loss: 0.6907 - val_acc: 0.5311 - val_auROC: 0.5924 - val_auPRC: 0.5600 - val_true_positives: 250.0000 - lr: 0.0010\n",
      "Epoch 7/100\n",
      "10/10 [==============================] - 2s 195ms/step - loss: 0.6913 - acc: 0.5226 - auROC: 0.5384 - auPRC: 0.5358 - true_positives: 453.0000 - val_loss: 0.6863 - val_acc: 0.5992 - val_auROC: 0.6269 - val_auPRC: 0.5946 - val_true_positives: 228.0000 - lr: 0.0010\n",
      "Epoch 8/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6897 - acc: 0.5374 - auROC: 0.5465 - auPRC: 0.5367 - true_positives: 441.0000 - val_loss: 0.6806 - val_acc: 0.5856 - val_auROC: 0.6470 - val_auPRC: 0.6242 - val_true_positives: 114.0000 - lr: 0.0010\n",
      "Epoch 9/100\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6871 - acc: 0.5516 - auROC: 0.5673 - auPRC: 0.5408 - true_positives: 610.0000 - val_loss: 0.6771 - val_acc: 0.5759 - val_auROC: 0.6486 - val_auPRC: 0.6267 - val_true_positives: 105.0000 - lr: 0.0010\n",
      "Epoch 10/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6773 - acc: 0.5568 - auROC: 0.6005 - auPRC: 0.5661 - true_positives: 494.0000 - val_loss: 0.6670 - val_acc: 0.6089 - val_auROC: 0.6475 - val_auPRC: 0.6262 - val_true_positives: 209.0000 - lr: 0.0010\n",
      "Epoch 11/100\n",
      "10/10 [==============================] - 2s 200ms/step - loss: 0.6733 - acc: 0.5789 - auROC: 0.6072 - auPRC: 0.5853 - true_positives: 620.0000 - val_loss: 0.6616 - val_acc: 0.6284 - val_auROC: 0.6498 - val_auPRC: 0.6195 - val_true_positives: 196.0000 - lr: 0.0010\n",
      "Epoch 12/100\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6662 - acc: 0.6011 - auROC: 0.6306 - auPRC: 0.6011 - true_positives: 608.0000 - val_loss: 0.6637 - val_acc: 0.6148 - val_auROC: 0.6481 - val_auPRC: 0.6210 - val_true_positives: 212.0000 - lr: 0.0010\n",
      "Epoch 13/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6646 - acc: 0.5863 - auROC: 0.6301 - auPRC: 0.6064 - true_positives: 618.0000 - val_loss: 0.6942 - val_acc: 0.5525 - val_auROC: 0.6458 - val_auPRC: 0.6119 - val_true_positives: 248.0000 - lr: 0.0010\n",
      "Epoch 14/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6712 - acc: 0.5868 - auROC: 0.6129 - auPRC: 0.5803 - true_positives: 553.0000 - val_loss: 0.6729 - val_acc: 0.6012 - val_auROC: 0.6379 - val_auPRC: 0.6065 - val_true_positives: 212.0000 - lr: 0.0010\n",
      "Epoch 15/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6618 - acc: 0.6053 - auROC: 0.6436 - auPRC: 0.6169 - true_positives: 669.0000 - val_loss: 0.6622 - val_acc: 0.5895 - val_auROC: 0.6608 - val_auPRC: 0.6352 - val_true_positives: 127.0000 - lr: 0.0010\n",
      "Epoch 16/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6553 - acc: 0.6158 - auROC: 0.6583 - auPRC: 0.6359 - true_positives: 608.0000 - val_loss: 0.6578 - val_acc: 0.6226 - val_auROC: 0.6580 - val_auPRC: 0.6254 - val_true_positives: 202.0000 - lr: 0.0010\n",
      "Epoch 17/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6489 - acc: 0.6232 - auROC: 0.6651 - auPRC: 0.6307 - true_positives: 642.0000 - val_loss: 0.6554 - val_acc: 0.6070 - val_auROC: 0.6595 - val_auPRC: 0.6342 - val_true_positives: 186.0000 - lr: 0.0010\n",
      "Epoch 18/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6433 - acc: 0.6384 - auROC: 0.6765 - auPRC: 0.6323 - true_positives: 653.0000 - val_loss: 0.6587 - val_acc: 0.6187 - val_auROC: 0.6654 - val_auPRC: 0.6341 - val_true_positives: 213.0000 - lr: 0.0010\n",
      "Epoch 19/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6519 - acc: 0.6200 - auROC: 0.6638 - auPRC: 0.6273 - true_positives: 644.0000 - val_loss: 0.6831 - val_acc: 0.5856 - val_auROC: 0.6620 - val_auPRC: 0.6441 - val_true_positives: 82.0000 - lr: 0.0010\n",
      "Epoch 20/100\n",
      "10/10 [==============================] - 2s 189ms/step - loss: 0.6603 - acc: 0.6126 - auROC: 0.6503 - auPRC: 0.6286 - true_positives: 646.0000 - val_loss: 0.6629 - val_acc: 0.5973 - val_auROC: 0.6496 - val_auPRC: 0.6164 - val_true_positives: 138.0000 - lr: 0.0010\n",
      "Epoch 21/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6406 - acc: 0.6311 - auROC: 0.6845 - auPRC: 0.6443 - true_positives: 548.0000 - val_loss: 0.6588 - val_acc: 0.6148 - val_auROC: 0.6618 - val_auPRC: 0.6311 - val_true_positives: 207.0000 - lr: 0.0010\n",
      "Epoch 22/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.6258 - acc: 0.6432 - auROC: 0.7037 - auPRC: 0.6721 - true_positives: 658.0000\n",
      "Epoch 22: ReduceLROnPlateau reducing learning rate to 0.00010000000474974513.\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6258 - acc: 0.6432 - auROC: 0.7037 - auPRC: 0.6721 - true_positives: 658.0000 - val_loss: 0.6566 - val_acc: 0.6148 - val_auROC: 0.6657 - val_auPRC: 0.6412 - val_true_positives: 186.0000 - lr: 0.0010\n",
      "Epoch 23/100\n",
      "10/10 [==============================] - 2s 195ms/step - loss: 0.6174 - acc: 0.6658 - auROC: 0.7167 - auPRC: 0.6801 - true_positives: 665.0000 - val_loss: 0.6562 - val_acc: 0.6148 - val_auROC: 0.6660 - val_auPRC: 0.6420 - val_true_positives: 184.0000 - lr: 1.0000e-04\n",
      "Epoch 24/100\n",
      "10/10 [==============================] - 2s 199ms/step - loss: 0.6128 - acc: 0.6695 - auROC: 0.7239 - auPRC: 0.6814 - true_positives: 645.0000 - val_loss: 0.6574 - val_acc: 0.6304 - val_auROC: 0.6654 - val_auPRC: 0.6401 - val_true_positives: 196.0000 - lr: 1.0000e-04\n",
      "Epoch 25/100\n",
      "10/10 [==============================] - 2s 193ms/step - loss: 0.6192 - acc: 0.6568 - auROC: 0.7175 - auPRC: 0.6839 - true_positives: 686.0000 - val_loss: 0.6585 - val_acc: 0.6265 - val_auROC: 0.6651 - val_auPRC: 0.6401 - val_true_positives: 201.0000 - lr: 1.0000e-04\n",
      "Epoch 26/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6160 - acc: 0.6595 - auROC: 0.7154 - auPRC: 0.6836 - true_positives: 671.0000 - val_loss: 0.6544 - val_acc: 0.6070 - val_auROC: 0.6685 - val_auPRC: 0.6459 - val_true_positives: 182.0000 - lr: 1.0000e-04\n",
      "Epoch 27/100\n",
      "10/10 [==============================] - 2s 200ms/step - loss: 0.6107 - acc: 0.6742 - auROC: 0.7258 - auPRC: 0.6936 - true_positives: 671.0000 - val_loss: 0.6557 - val_acc: 0.6304 - val_auROC: 0.6680 - val_auPRC: 0.6452 - val_true_positives: 196.0000 - lr: 1.0000e-04\n",
      "Epoch 28/100\n",
      "10/10 [==============================] - 2s 200ms/step - loss: 0.6088 - acc: 0.6679 - auROC: 0.7281 - auPRC: 0.6982 - true_positives: 661.0000 - val_loss: 0.6547 - val_acc: 0.6187 - val_auROC: 0.6691 - val_auPRC: 0.6458 - val_true_positives: 189.0000 - lr: 1.0000e-04\n",
      "Epoch 29/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6123 - acc: 0.6742 - auROC: 0.7240 - auPRC: 0.6834 - true_positives: 669.0000 - val_loss: 0.6547 - val_acc: 0.6265 - val_auROC: 0.6690 - val_auPRC: 0.6474 - val_true_positives: 193.0000 - lr: 1.0000e-04\n",
      "Epoch 30/100\n",
      "10/10 [==============================] - 2s 195ms/step - loss: 0.6109 - acc: 0.6737 - auROC: 0.7256 - auPRC: 0.6966 - true_positives: 677.0000 - val_loss: 0.6548 - val_acc: 0.6245 - val_auROC: 0.6695 - val_auPRC: 0.6462 - val_true_positives: 194.0000 - lr: 1.0000e-04\n",
      "Epoch 31/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.6081 - acc: 0.6768 - auROC: 0.7327 - auPRC: 0.7016 - true_positives: 685.0000\n",
      "Epoch 31: ReduceLROnPlateau reducing learning rate to 1.0000000474974514e-05.\n",
      "10/10 [==============================] - 2s 199ms/step - loss: 0.6081 - acc: 0.6768 - auROC: 0.7327 - auPRC: 0.7016 - true_positives: 685.0000 - val_loss: 0.6544 - val_acc: 0.6226 - val_auROC: 0.6699 - val_auPRC: 0.6476 - val_true_positives: 193.0000 - lr: 1.0000e-04\n",
      "Epoch 32/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6101 - acc: 0.6595 - auROC: 0.7274 - auPRC: 0.7015 - true_positives: 655.0000 - val_loss: 0.6542 - val_acc: 0.6226 - val_auROC: 0.6707 - val_auPRC: 0.6485 - val_true_positives: 192.0000 - lr: 1.0000e-05\n",
      "Epoch 33/100\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6029 - acc: 0.6795 - auROC: 0.7371 - auPRC: 0.7082 - true_positives: 679.0000 - val_loss: 0.6538 - val_acc: 0.6245 - val_auROC: 0.6705 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-05\n",
      "Epoch 34/100\n",
      "10/10 [==============================] - 2s 198ms/step - loss: 0.6086 - acc: 0.6732 - auROC: 0.7284 - auPRC: 0.7075 - true_positives: 682.0000 - val_loss: 0.6537 - val_acc: 0.6226 - val_auROC: 0.6705 - val_auPRC: 0.6473 - val_true_positives: 189.0000 - lr: 1.0000e-05\n",
      "Epoch 35/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6065 - acc: 0.6705 - auROC: 0.7303 - auPRC: 0.6881 - true_positives: 679.0000 - val_loss: 0.6538 - val_acc: 0.6245 - val_auROC: 0.6703 - val_auPRC: 0.6480 - val_true_positives: 190.0000 - lr: 1.0000e-05\n",
      "Epoch 36/100\n",
      "10/10 [==============================] - 2s 193ms/step - loss: 0.6200 - acc: 0.6679 - auROC: 0.7120 - auPRC: 0.6780 - true_positives: 672.0000 - val_loss: 0.6538 - val_acc: 0.6265 - val_auROC: 0.6704 - val_auPRC: 0.6476 - val_true_positives: 191.0000 - lr: 1.0000e-05\n",
      "Epoch 37/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6040 - acc: 0.6726 - auROC: 0.7317 - auPRC: 0.7043 - true_positives: 666.0000 - val_loss: 0.6537 - val_acc: 0.6245 - val_auROC: 0.6703 - val_auPRC: 0.6478 - val_true_positives: 190.0000 - lr: 1.0000e-05\n",
      "Epoch 38/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6032 - acc: 0.6737 - auROC: 0.7344 - auPRC: 0.6882 - true_positives: 665.0000 - val_loss: 0.6540 - val_acc: 0.6245 - val_auROC: 0.6702 - val_auPRC: 0.6476 - val_true_positives: 191.0000 - lr: 1.0000e-05\n",
      "Epoch 39/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.6098 - acc: 0.6658 - auROC: 0.7252 - auPRC: 0.6868 - true_positives: 674.0000\n",
      "Epoch 39: ReduceLROnPlateau reducing learning rate to 1.0000000656873453e-06.\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6098 - acc: 0.6658 - auROC: 0.7252 - auPRC: 0.6868 - true_positives: 674.0000 - val_loss: 0.6539 - val_acc: 0.6245 - val_auROC: 0.6702 - val_auPRC: 0.6478 - val_true_positives: 191.0000 - lr: 1.0000e-05\n",
      "Epoch 40/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6067 - acc: 0.6742 - auROC: 0.7316 - auPRC: 0.6963 - true_positives: 671.0000 - val_loss: 0.6540 - val_acc: 0.6245 - val_auROC: 0.6702 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-06\n",
      "Epoch 41/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6075 - acc: 0.6721 - auROC: 0.7324 - auPRC: 0.7025 - true_positives: 679.0000 - val_loss: 0.6540 - val_acc: 0.6245 - val_auROC: 0.6704 - val_auPRC: 0.6479 - val_true_positives: 191.0000 - lr: 1.0000e-06\n",
      "Epoch 42/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6121 - acc: 0.6611 - auROC: 0.7228 - auPRC: 0.6898 - true_positives: 667.0000 - val_loss: 0.6540 - val_acc: 0.6245 - val_auROC: 0.6705 - val_auPRC: 0.6479 - val_true_positives: 191.0000 - lr: 1.0000e-06\n",
      "Epoch 43/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6045 - acc: 0.6805 - auROC: 0.7353 - auPRC: 0.6994 - true_positives: 693.0000 - val_loss: 0.6539 - val_acc: 0.6245 - val_auROC: 0.6700 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-06\n",
      "Epoch 44/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.5966 - acc: 0.6768 - auROC: 0.7427 - auPRC: 0.7090 - true_positives: 671.0000\n",
      "Epoch 44: ReduceLROnPlateau reducing learning rate to 1.0000001111620805e-07.\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.5966 - acc: 0.6768 - auROC: 0.7427 - auPRC: 0.7090 - true_positives: 671.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-06\n",
      "Epoch 45/100\n",
      "10/10 [==============================] - 2s 190ms/step - loss: 0.6103 - acc: 0.6668 - auROC: 0.7267 - auPRC: 0.6854 - true_positives: 672.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6475 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 46/100\n",
      "10/10 [==============================] - 2s 197ms/step - loss: 0.6032 - acc: 0.6747 - auROC: 0.7343 - auPRC: 0.6980 - true_positives: 680.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6475 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 47/100\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6091 - acc: 0.6721 - auROC: 0.7321 - auPRC: 0.7133 - true_positives: 674.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 48/100\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6026 - acc: 0.6784 - auROC: 0.7380 - auPRC: 0.7078 - true_positives: 683.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6475 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 49/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.6035 - acc: 0.6695 - auROC: 0.7348 - auPRC: 0.7114 - true_positives: 679.0000\n",
      "Epoch 49: ReduceLROnPlateau reducing learning rate to 1e-07.\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6035 - acc: 0.6695 - auROC: 0.7348 - auPRC: 0.7114 - true_positives: 679.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6702 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 50/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6168 - acc: 0.6532 - auROC: 0.7135 - auPRC: 0.6806 - true_positives: 652.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6702 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 51/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6037 - acc: 0.6711 - auROC: 0.7355 - auPRC: 0.6875 - true_positives: 675.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6702 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 52/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6061 - acc: 0.6784 - auROC: 0.7325 - auPRC: 0.7037 - true_positives: 675.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6702 - val_auPRC: 0.6477 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 53/100\n",
      "10/10 [==============================] - 2s 194ms/step - loss: 0.6036 - acc: 0.6721 - auROC: 0.7360 - auPRC: 0.6995 - true_positives: 676.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 54/100\n",
      "10/10 [==============================] - 2s 184ms/step - loss: 0.6091 - acc: 0.6695 - auROC: 0.7291 - auPRC: 0.6951 - true_positives: 673.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 55/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6117 - acc: 0.6658 - auROC: 0.7254 - auPRC: 0.6970 - true_positives: 662.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 56/100\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6064 - acc: 0.6695 - auROC: 0.7309 - auPRC: 0.7003 - true_positives: 662.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 57/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6015 - acc: 0.6732 - auROC: 0.7368 - auPRC: 0.7062 - true_positives: 665.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 58/100\n",
      "10/10 [==============================] - 2s 189ms/step - loss: 0.6011 - acc: 0.6758 - auROC: 0.7368 - auPRC: 0.7036 - true_positives: 682.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 59/100\n",
      "10/10 [==============================] - 2s 189ms/step - loss: 0.6093 - acc: 0.6795 - auROC: 0.7301 - auPRC: 0.7020 - true_positives: 688.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 60/100\n",
      "10/10 [==============================] - 2s 196ms/step - loss: 0.6086 - acc: 0.6763 - auROC: 0.7274 - auPRC: 0.6868 - true_positives: 686.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 61/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6098 - acc: 0.6832 - auROC: 0.7311 - auPRC: 0.6903 - true_positives: 700.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 62/100\n",
      "10/10 [==============================] - 2s 193ms/step - loss: 0.6036 - acc: 0.6695 - auROC: 0.7323 - auPRC: 0.7070 - true_positives: 666.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6473 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 63/100\n",
      "10/10 [==============================] - 2s 185ms/step - loss: 0.6124 - acc: 0.6574 - auROC: 0.7237 - auPRC: 0.6860 - true_positives: 674.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 64/100\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6036 - acc: 0.6726 - auROC: 0.7362 - auPRC: 0.7027 - true_positives: 687.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 65/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6064 - acc: 0.6805 - auROC: 0.7316 - auPRC: 0.7075 - true_positives: 673.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 66/100\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6022 - acc: 0.6658 - auROC: 0.7355 - auPRC: 0.7027 - true_positives: 663.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 67/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6022 - acc: 0.6847 - auROC: 0.7380 - auPRC: 0.7028 - true_positives: 680.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 68/100\n",
      "10/10 [==============================] - 2s 188ms/step - loss: 0.6125 - acc: 0.6721 - auROC: 0.7196 - auPRC: 0.6819 - true_positives: 678.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6475 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 69/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6072 - acc: 0.6700 - auROC: 0.7291 - auPRC: 0.6970 - true_positives: 682.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6475 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 70/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6074 - acc: 0.6774 - auROC: 0.7328 - auPRC: 0.7019 - true_positives: 687.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 71/100\n",
      "10/10 [==============================] - 2s 187ms/step - loss: 0.6164 - acc: 0.6584 - auROC: 0.7172 - auPRC: 0.6822 - true_positives: 657.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6701 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 72/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6035 - acc: 0.6737 - auROC: 0.7382 - auPRC: 0.7035 - true_positives: 666.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 73/100\n",
      "10/10 [==============================] - 2s 191ms/step - loss: 0.6103 - acc: 0.6563 - auROC: 0.7234 - auPRC: 0.6982 - true_positives: 654.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 74/100\n",
      "10/10 [==============================] - ETA: 0s - loss: 0.6129 - acc: 0.6716 - auROC: 0.7215 - auPRC: 0.6883 - true_positives: 675.0000Restoring model weights from the end of the best epoch: 34.\n",
      "10/10 [==============================] - 2s 192ms/step - loss: 0.6129 - acc: 0.6716 - auROC: 0.7215 - auPRC: 0.6883 - true_positives: 675.0000 - val_loss: 0.6539 - val_acc: 0.6265 - val_auROC: 0.6700 - val_auPRC: 0.6474 - val_true_positives: 191.0000 - lr: 1.0000e-07\n",
      "Epoch 74: early stopping\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "history = model.fit(train_data_gen,\n",
    "\t\t\t\t\tsteps_per_epoch=math.ceil(\n",
    "\t\t\t\t\t\ttrain_df.shape[0] / BATCH_SIZE),\n",
    "\t\t\t\t\tepochs=100,\n",
    "\t\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\t\tcallbacks=callbacks\n",
    "\t\t\t\t\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: Loss of 0.6536747217178345; accuracy of 0.6225680708885193; auroc of 0.6704593896865845; auprc of 0.6473498344421387; TP of 189.0\n"
     ]
    }
   ],
   "source": [
    "scores = model.evaluate(x_validation, y_validation, verbose=0)\n",
    "print(f'Score: Loss of {scores[0]}; accuracy of {scores[1]}; auroc of {scores[2]}; auprc of {scores[3]}; TP of {scores[4]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finetune phylo-aug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train_pure, y_train_pure = get_input_and_labels(train_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38/38 [==============================] - 2s 24ms/step - loss: 0.6128 - acc: 0.6582 - auROC: 0.7219 - auPRC: 0.6870 - true_positives_1: 1325.0000 - val_loss: 0.6559 - val_acc: 0.6245 - val_auROC: 0.6697 - val_auPRC: 0.6446 - val_true_positives_1: 200.0000\n",
      "Epoch 2/10\n",
      "38/38 [==============================] - 1s 16ms/step - loss: 0.6096 - acc: 0.6682 - auROC: 0.7283 - auPRC: 0.6935 - true_positives_1: 1348.0000 - val_loss: 0.6535 - val_acc: 0.6245 - val_auROC: 0.6727 - val_auPRC: 0.6473 - val_true_positives_1: 196.0000\n",
      "Epoch 3/10\n",
      "38/38 [==============================] - 1s 16ms/step - loss: 0.6071 - acc: 0.6671 - auROC: 0.7296 - auPRC: 0.7003 - true_positives_1: 1340.0000 - val_loss: 0.6513 - val_acc: 0.6265 - val_auROC: 0.6758 - val_auPRC: 0.6513 - val_true_positives_1: 188.0000\n",
      "Epoch 4/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.6024 - acc: 0.6750 - auROC: 0.7371 - auPRC: 0.7040 - true_positives_1: 1360.0000 - val_loss: 0.6502 - val_acc: 0.6284 - val_auROC: 0.6774 - val_auPRC: 0.6560 - val_true_positives_1: 188.0000\n",
      "Epoch 5/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.6011 - acc: 0.6742 - auROC: 0.7363 - auPRC: 0.7034 - true_positives_1: 1318.0000 - val_loss: 0.6534 - val_acc: 0.6265 - val_auROC: 0.6774 - val_auPRC: 0.6543 - val_true_positives_1: 202.0000\n",
      "Epoch 6/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.5996 - acc: 0.6813 - auROC: 0.7422 - auPRC: 0.7106 - true_positives_1: 1380.0000 - val_loss: 0.6504 - val_acc: 0.6401 - val_auROC: 0.6787 - val_auPRC: 0.6581 - val_true_positives_1: 198.0000\n",
      "Epoch 7/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.5914 - acc: 0.6834 - auROC: 0.7495 - auPRC: 0.7176 - true_positives_1: 1357.0000 - val_loss: 0.6497 - val_acc: 0.6381 - val_auROC: 0.6805 - val_auPRC: 0.6621 - val_true_positives_1: 195.0000\n",
      "Epoch 8/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.5894 - acc: 0.6921 - auROC: 0.7536 - auPRC: 0.7209 - true_positives_1: 1386.0000 - val_loss: 0.6496 - val_acc: 0.6362 - val_auROC: 0.6814 - val_auPRC: 0.6633 - val_true_positives_1: 195.0000\n",
      "Epoch 9/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.5814 - acc: 0.6958 - auROC: 0.7609 - auPRC: 0.7237 - true_positives_1: 1391.0000 - val_loss: 0.6492 - val_acc: 0.6284 - val_auROC: 0.6842 - val_auPRC: 0.6680 - val_true_positives_1: 174.0000\n",
      "Epoch 10/10\n",
      "38/38 [==============================] - 1s 15ms/step - loss: 0.5833 - acc: 0.7003 - auROC: 0.7582 - auPRC: 0.7219 - true_positives_1: 1407.0000 - val_loss: 0.6491 - val_acc: 0.6323 - val_auROC: 0.6832 - val_auPRC: 0.6653 - val_true_positives_1: 172.0000\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<keras.src.callbacks.History at 0x7fea49fbb760>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tensorflow import keras\n",
    "from src.diff_expression_model import compile_model\n",
    "\n",
    "# set up callbacks\n",
    "model = compile_model(model, learning_rate=0.0001)\n",
    "\n",
    "# train model\n",
    "model.fit(x_train_pure, y_train_pure,\n",
    "\t\t\t\tepochs=10,\n",
    "\t\t\t\tbatch_size=100,\n",
    "\t\t\t\tshuffle=True,\n",
    "\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\tcallbacks=[early_stopping_callback])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: Loss of 0.6490797996520996; accuracy of 0.6322957277297974; auroc of 0.6831925511360168; auprc of 0.6652507781982422; TP of 172.0\n"
     ]
    }
   ],
   "source": [
    "scores = model.evaluate(x_validation, y_validation, verbose=0)\n",
    "print(f'Score: Loss of {scores[0]}; accuracy of {scores[1]}; auroc of {scores[2]}; auprc of {scores[3]}; TP of {scores[4]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"/home/ubuntu/DeepDifE/data/phylo_aug/results/result_phylo_aug_finetune.txt\", \"a\")\n",
    "f.write(f'Scores: Max_phylo\\n')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We perform post-hoc conjoing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "siamese_model = get_siamese_model(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare the test set data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test = np.stack(test_df[\"One_hot_encoded\"])\n",
    "x_test_rc = np.stack(test_df[\"RC_one_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_test = test_df[\"Label\"].to_numpy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we used evo aug in our model all the sequences which were trained on, were of length 620. We use the evoaug padding function to pad the test set to the same length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test = model._pad_end(x_test)\n",
    "x_test_rc = model._pad_end(x_test_rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8/8 [==============================] - 0s 3ms/step\n"
     ]
    }
   ],
   "source": [
    "predictions_categories, predictions = post_hoc_conjoining(siamese_model, x_test, x_test_rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7221760221760222"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_auroc(y_test, predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.64      0.64      0.64       111\n",
      "           1       0.69      0.69      0.69       130\n",
      "\n",
      "    accuracy                           0.67       241\n",
      "   macro avg       0.67      0.67      0.67       241\n",
      "weighted avg       0.67      0.67      0.67       241\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(y_test, predictions_categories))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
