{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/DeepDifE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-09 11:19:25.461706: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-08-09 11:19:25.504893: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2024-08-09 11:19:25.506005: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2024-08-09 11:19:26.226815: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%cd ../..\n",
    "\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from evoaug_tf import evoaug, augment\n",
    "from src.diff_expression_model import get_model, get_siamese_model, post_hoc_conjoining, get_auroc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now we start from this data pickle as I'm not aware how Helder did DE analysis and generated the labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppath = \"data/dataset_solid_chrome.pkl\"\n",
    "with open(ppath, 'rb') as f:\n",
    "    data = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>seqs</th>\n",
       "      <th>ohs</th>\n",
       "      <th>rcohs</th>\n",
       "      <th>ohsDuo</th>\n",
       "      <th>in_original_balanced</th>\n",
       "      <th>set</th>\n",
       "      <th>npshap-single</th>\n",
       "      <th>npshap-posthoc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AT4G27120</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[-0.00029412], [0.00011191], [0.00071743], ...</td>\n",
       "      <td>[[[[-0.00024759], [0.00013361], [0.00015087], ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT4G19600</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[0.00049611], [0.00069235], [0.00377408], [...</td>\n",
       "      <td>[[[[0.00051932], [0.00032187], [0.00027891], [...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT3G60880</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[-0.00030243], [-1.08283739e-05], [0.000359...</td>\n",
       "      <td>[[[[-0.00031496], [-8.34961975e-05], [0.000132...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G06960</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[0.00254294], [0.00639127], [0.00152307], [...</td>\n",
       "      <td>[[[[0.00208253], [0.0018626], [0.00023493], [0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G14890</th>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>valid</td>\n",
       "      <td>[[[[-5.23597009e-05], [0.00045373], [0.0003575...</td>\n",
       "      <td>[[[[0.00051831], [0.00050734], [4.30682958e-05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G64230</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[[[[-1.85048248e-05], [5.31432997e-05], [0.000...</td>\n",
       "      <td>[[[[0.00034249], [0.00039158], [1.82568183e-05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT5G64780</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[9.53665585e-05], [0.00032812], [0.00046541...</td>\n",
       "      <td>[[[[-5.87131144e-05], [6.68894427e-05], [6.172...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT4G30470</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>test</td>\n",
       "      <td>[[[[0.00061825], [6.60565856e-05], [5.89712399...</td>\n",
       "      <td>[[[[0.00044855], [6.3632986e-05], [0.00018469]...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT3G51895</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[-0.00021527], [0.00037615], [0.0002829], [...</td>\n",
       "      <td>[[[[-0.00010363], [0.00015631], [0.0001], [0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT2G35585</th>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...</td>\n",
       "      <td>True</td>\n",
       "      <td>train</td>\n",
       "      <td>[[[[0.00069553], [0.00023614], [0.00080674], [...</td>\n",
       "      <td>[[[[0.00047114], [9.892659e-05], [0.00010585],...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Category    GeneFamily  \\\n",
       "geneID                             \n",
       "AT4G27120        0  HOM04D000881   \n",
       "AT4G19600        0  HOM04D000740   \n",
       "AT3G60880        0  HOM04D003119   \n",
       "AT5G06960        0  HOM04D000319   \n",
       "AT1G14890        0  HOM04D000273   \n",
       "...            ...           ...   \n",
       "AT5G64230        1  HOM04D003278   \n",
       "AT5G64780        1  HOM04D002552   \n",
       "AT4G30470        1  HOM04D000082   \n",
       "AT3G51895        1  HOM04D000270   \n",
       "AT2G35585        1  HOM04D004118   \n",
       "\n",
       "                                                        seqs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "AT4G19600  GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "AT3G60880  AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "AT5G06960  CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "AT1G14890  TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                      ...   \n",
       "AT5G64230  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "AT5G64780  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "AT4G30470  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "AT3G51895  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "AT2G35585  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                                         ohs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT4G19600  [[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...   \n",
       "AT3G60880  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "AT5G06960  [[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "AT1G14890  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT5G64780  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "AT4G30470  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "AT3G51895  [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT2G35585  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "\n",
       "                                                       rcohs  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT4G19600  [[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT3G60880  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT5G06960  [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "AT1G14890  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "AT5G64780  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...   \n",
       "AT4G30470  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "AT3G51895  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "AT2G35585  [[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...   \n",
       "\n",
       "                                                      ohsDuo  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1...   \n",
       "AT4G19600  [[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1...   \n",
       "AT3G60880  [[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0...   \n",
       "AT5G06960  [[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0...   \n",
       "AT1G14890  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1...   \n",
       "AT5G64780  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...   \n",
       "AT4G30470  [[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0...   \n",
       "AT3G51895  [[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0...   \n",
       "AT2G35585  [[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0...   \n",
       "\n",
       "           in_original_balanced    set  \\\n",
       "geneID                                   \n",
       "AT4G27120                  True  train   \n",
       "AT4G19600                  True  valid   \n",
       "AT3G60880                  True  valid   \n",
       "AT5G06960                  True  train   \n",
       "AT1G14890                  True  valid   \n",
       "...                         ...    ...   \n",
       "AT5G64230                  True   test   \n",
       "AT5G64780                  True  train   \n",
       "AT4G30470                  True   test   \n",
       "AT3G51895                  True  train   \n",
       "AT2G35585                  True  train   \n",
       "\n",
       "                                               npshap-single  \\\n",
       "geneID                                                         \n",
       "AT4G27120  [[[[-0.00029412], [0.00011191], [0.00071743], ...   \n",
       "AT4G19600  [[[[0.00049611], [0.00069235], [0.00377408], [...   \n",
       "AT3G60880  [[[[-0.00030243], [-1.08283739e-05], [0.000359...   \n",
       "AT5G06960  [[[[0.00254294], [0.00639127], [0.00152307], [...   \n",
       "AT1G14890  [[[[-5.23597009e-05], [0.00045373], [0.0003575...   \n",
       "...                                                      ...   \n",
       "AT5G64230  [[[[-1.85048248e-05], [5.31432997e-05], [0.000...   \n",
       "AT5G64780  [[[[9.53665585e-05], [0.00032812], [0.00046541...   \n",
       "AT4G30470  [[[[0.00061825], [6.60565856e-05], [5.89712399...   \n",
       "AT3G51895  [[[[-0.00021527], [0.00037615], [0.0002829], [...   \n",
       "AT2G35585  [[[[0.00069553], [0.00023614], [0.00080674], [...   \n",
       "\n",
       "                                              npshap-posthoc  \n",
       "geneID                                                        \n",
       "AT4G27120  [[[[-0.00024759], [0.00013361], [0.00015087], ...  \n",
       "AT4G19600  [[[[0.00051932], [0.00032187], [0.00027891], [...  \n",
       "AT3G60880  [[[[-0.00031496], [-8.34961975e-05], [0.000132...  \n",
       "AT5G06960  [[[[0.00208253], [0.0018626], [0.00023493], [0...  \n",
       "AT1G14890  [[[[0.00051831], [0.00050734], [4.30682958e-05...  \n",
       "...                                                      ...  \n",
       "AT5G64230  [[[[0.00034249], [0.00039158], [1.82568183e-05...  \n",
       "AT5G64780  [[[[-5.87131144e-05], [6.68894427e-05], [6.172...  \n",
       "AT4G30470  [[[[0.00044855], [6.3632986e-05], [0.00018469]...  \n",
       "AT3G51895  [[[[-0.00010363], [0.00015631], [0.0001], [0.0...  \n",
       "AT2G35585  [[[[0.00047114], [9.892659e-05], [0.00010585],...  \n",
       "\n",
       "[2398 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "train    1900\n",
       "valid     257\n",
       "test      241\n",
       "Name: set, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"set\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To show how to subdivide the dataset into train-test split we only take the following columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = data.reset_index()\n",
    "dataset = dataset[[\"geneID\", \"Category\", \"GeneFamily\", \"seqs\"]]\n",
    "dataset.rename(columns={\"geneID\":\"GeneID\", \"Category\":\"Label\", \"seqs\": \"Sequence\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylo augmentation "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the fasta files for each species of which we want to use orthologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_augmentation_base_path = \"data/phylo_aug/fasta/\"\n",
    "species_fasta_files = {\n",
    "\t\t\t\"aly\":\"aly.tss.fasta\",\n",
    "\t\t\t\"chi\":\"chi.tss.fasta\",\n",
    "\t\t\t\"cru\":\"cru.tss.fasta\",\n",
    "\t\t\t\"aar\":\"aar.tss.fasta\",\n",
    "\t\t\t\"ath-an1\":\"ath-an1.tss.fasta\",\n",
    "\t\t\t\"ath-c24\":\"ath-c24.tss.fasta\",\n",
    "\t\t\t\"ath-cvi\":\"ath-cvi.tss.fasta\",\n",
    "\t\t\t\"ath-eri\":\"ath-eri.tss.fasta\",\n",
    "\t\t\t\"ath-kyo\":\"ath-kyo.tss.fasta\",\n",
    "\t\t\t\"ath-ler\":\"ath-ler.tss.fasta\",\n",
    "\t\t\t\"ath-sha\":\"ath-sha.tss.fasta\",\n",
    "\t\t\t\"ath\":\"ath.tss.fasta\",\n",
    "\t\t\t\"chi\":\"chi.tss.fasta\",\n",
    "\t\t\t\"cpa\":\"cpa.tss.fasta\",\n",
    "\t\t\t\"cru\":\"cru.tss.fasta\",\n",
    "\t\t\t\"esa\":\"esa.tss.fasta\",\n",
    "\t\t\t\"spa\":\"spa.tss.fasta\",\n",
    "\t\t\t\"tha\":\"tha.tss.fasta\"\n",
    "}\n",
    "\n",
    "species_to_include = [\"aly\",\"chi\",\"cru\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsing the fasta files. In this case the label of the original sequence is incorporated in the ids in the fasta file\n",
    "e.g:\n",
    "```\n",
    ">Aa31LG6G5540_1\n",
    "TTGTATAAATATATTAAATT\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: data/phylo_aug/fasta/aly.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/chi.tss.fasta done!\n",
      "File: data/phylo_aug/fasta/cru.tss.fasta done!\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequence_list = []\n",
    "gene_id_list = []\n",
    "label_list = []\n",
    "species_list = []\n",
    "for species in species_to_include:\n",
    "\tspecies_fasta = species_fasta_files[species]\n",
    "\tfasta_path = f\"{phylo_augmentation_base_path}{species_fasta}\"\n",
    "\tfasta_sequences = SeqIO.parse(open(fasta_path),'fasta')\n",
    "\tprint(f\"File: {fasta_path} done!\")\n",
    "\tfor fasta in fasta_sequences:\n",
    "\t\tname, sequence = fasta.id, str(fasta.seq)\n",
    "\t\t\n",
    "\t\t# The label is stored inside the id name (e.g. >ATERI-4G42060_1), split these\n",
    "\t\tgene_id = name.split(\"_\")[0]\n",
    "\t\tlabel = name.split(\"_\")[1]\n",
    "\t\tspecies = species_fasta.split(\".\")[0]\n",
    "\n",
    "\t\tgene_id_list.append(gene_id)\n",
    "\t\tsequence_list.append(sequence)\n",
    "\t\tlabel_list.append(label)\n",
    "\t\tspecies_list.append(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.DataFrame({\"GeneID\": gene_id_list, \"Label\": label_list, \"Sequence\": sequence_list, \"Species\": species_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  \n",
       "...                                                 ...     ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  \n",
       "\n",
       "[2794 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get gene family of the original ortholog"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because during the batch time augmentation a original sample will be swapped by one of its orthologs, we need to know from which original sample each ortholog originates from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "base_dir_ortholog = \"data/phylo_aug/orthologs/\"\n",
    "ortholog_files = {\n",
    "\t\t\t\"ath-aly\":\"ath_aly.txt\",\n",
    "\t\t\t\"ath-chi\":\"ath_chi.txt\",\n",
    "\t\t\t\"ath-cru\":\"ath_cru.txt\",\n",
    "\t\t\t\"ath-aar\":\"ath_aar.txt\",\n",
    "\t\t\t\"ath-aly\":\"ath_aly.txt\",\n",
    "\t\t\t\"ath-ath-an1\":\"ath_ath-an1.txt\",\n",
    "\t\t\t\"ath-ath-c24\":\"ath_ath-c24.txt\",\n",
    "\t\t\t\"ath-ath-cvi\":\"ath_ath-cvi.txt\",\n",
    "\t\t\t\"ath-ath-eri\":\"ath_ath-eri.txt\",\n",
    "\t\t\t\"ath-ath-kyo\":\"ath_ath-kyo.txt\",\n",
    "\t\t\t\"ath-ath-ler\":\"ath_ath-ler.txt\",\n",
    "\t\t\t\"ath-ath-sha\":\"ath_ath-sha.txt\",\n",
    "\t\t\t\"ath-ath\":\"ath_ath.txt\",\n",
    "\t\t\t\"ath-chi\":\"ath_chi.txt\",\n",
    "\t\t\t\"ath-cpa\":\"ath_cpa.txt\",\n",
    "\t\t\t\"ath-cru\":\"ath_cru.txt\",\n",
    "\t\t\t\"ath-esa\":\"ath_esa.txt\",\n",
    "\t\t\t\"ath-spa\":\"ath_spa.txt\",\n",
    "\t\t\t\"ath-tha\":\"ath_tha.txt\"\n",
    "\t\t\t}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a lookup table to include the orthologs in the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "lookup_table = {}\n",
    "for orthologs in ortholog_files.values():\n",
    "\twith open(f\"{base_dir_ortholog}{orthologs}\", 'r') as file:\n",
    "\t\tfor line in file:\n",
    "\t\t\t# Split each line by tab\n",
    "\t\t\tcolumns = line.strip().split('\\t')\n",
    "\t\t\tif len(columns) >= 2:  # Ensure there are at least two columns\n",
    "\t\t\t\tkey = columns[1]\n",
    "\t\t\t\tvalue = columns[0]\n",
    "\t\t\t\tlookup_table[key] = value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df[\"Ortholog\"] = phylo_aug_df[\"GeneID\"].map(lookup_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720  \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010  \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340  \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820  \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260  \n",
       "...                                                 ...     ...        ...  \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740  \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910  \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460  \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530  \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970  \n",
       "\n",
       "[2794 rows x 5 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylo_aug_df = pd.merge(phylo_aug_df, dataset[[\"GeneID\", \"GeneFamily\"]], left_on=\"Ortholog\", right_on=\"GeneID\", how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily  \n",
       "0     HOM04D003099  \n",
       "1     HOM04D003060  \n",
       "2     HOM04D000135  \n",
       "3     HOM04D001302  \n",
       "4     HOM04D006130  \n",
       "...            ...  \n",
       "2789  HOM04D004133  \n",
       "2790  HOM04D001924  \n",
       "2791  HOM04D006357  \n",
       "2792  HOM04D007279  \n",
       "2793  HOM04D001540  \n",
       "\n",
       "[2794 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df.drop(\"GeneID_y\", axis=1, inplace=True)\n",
    "phylo_aug_df.rename(columns={\"GeneID_x\": \"GeneID\"}, inplace=True)\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One-hot-encode & reverse-complement both dataset as augmented dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import one_hot_encode_series, reverse_complement_series, reverse_complement_sequence\n",
    "dataset[\"One_hot_encoded\"] = one_hot_encode_series(dataset[\"Sequence\"])\n",
    "phylo_aug_df[\"One_hot_encoded\"] = one_hot_encode_series(phylo_aug_df[\"Sequence\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset[\"RC_one_hot_encoded\"] = reverse_complement_series(dataset[\"One_hot_encoded\"])\n",
    "phylo_aug_df[\"RC_one_hot_encoded\"] = reverse_complement_series(phylo_aug_df[\"One_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT4G27120</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000881</td>\n",
       "      <td>TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G19600</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000740</td>\n",
       "      <td>GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT3G60880</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D003119</td>\n",
       "      <td>AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G06960</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000319</td>\n",
       "      <td>CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...</td>\n",
       "      <td>[[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G14890</td>\n",
       "      <td>0</td>\n",
       "      <td>HOM04D000273</td>\n",
       "      <td>TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2393</th>\n",
       "      <td>AT5G64230</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D003278</td>\n",
       "      <td>AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>AT5G64780</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D002552</td>\n",
       "      <td>TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2395</th>\n",
       "      <td>AT4G30470</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000082</td>\n",
       "      <td>TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2396</th>\n",
       "      <td>AT3G51895</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D000270</td>\n",
       "      <td>TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2397</th>\n",
       "      <td>AT2G35585</td>\n",
       "      <td>1</td>\n",
       "      <td>HOM04D004118</td>\n",
       "      <td>TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2398 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GeneID Label    GeneFamily  \\\n",
       "0     AT4G27120     0  HOM04D000881   \n",
       "1     AT4G19600     0  HOM04D000740   \n",
       "2     AT3G60880     0  HOM04D003119   \n",
       "3     AT5G06960     0  HOM04D000319   \n",
       "4     AT1G14890     0  HOM04D000273   \n",
       "...         ...   ...           ...   \n",
       "2393  AT5G64230     1  HOM04D003278   \n",
       "2394  AT5G64780     1  HOM04D002552   \n",
       "2395  AT4G30470     1  HOM04D000082   \n",
       "2396  AT3G51895     1  HOM04D000270   \n",
       "2397  AT2G35585     1  HOM04D004118   \n",
       "\n",
       "                                               Sequence  \\\n",
       "0     TAGAGAAGACAAGCGGTTATTTCGTAATTTCCCAGCGACTTTGAAA...   \n",
       "1     GTCAAGTAGTGAAATCAAGGTGTGAAGTAAGCTGAGGACAGATAAT...   \n",
       "2     AGTTGATATTGAATGAAATCTTCATGTTTTTTGATAAATGATTATA...   \n",
       "3     CACTTGTCAGATTCTTCTTACCAAATCCATCAACAAATAAGCAAAT...   \n",
       "4     TTGATATAACAGATTCAACACTAAAAATGAGTAAAATCTAAAAAAG...   \n",
       "...                                                 ...   \n",
       "2393  AAGAAAGAAAAACCGTACATAAACACCCATCTGGTATACCATCGTC...   \n",
       "2394  TTTTAGAAAGAAGAAGAAGGATTATTGCCTTATTGGTGAAGGGAAG...   \n",
       "2395  TATGTACAGTCTCTACATTTTTTCAAATACATTTTTTTCTTTTTCA...   \n",
       "2396  TGGTAAATAATTAAATATATAAGAACATTATTCTAAAGCGTTGAAT...   \n",
       "2397  TTTTGCCACTGCTGGCTCTGATGGTGCTTTCAATTTCTAGGACAAG...   \n",
       "\n",
       "                                        One_hot_encoded  \\\n",
       "0     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "1     [[0, 0, 1, 0], [0, 0, 0, 1], [0, 1, 0, 0], [1,...   \n",
       "2     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...   \n",
       "3     [[0, 1, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "4     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...   \n",
       "...                                                 ...   \n",
       "2393  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 1, 0], [1,...   \n",
       "2394  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "2395  [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2396  [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "2397  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "1     [[0, 0, 1, 0], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2     [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "3     [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [1,...  \n",
       "4     [[0, 0, 0, 1], [1, 0, 0, 0], [0, 0, 0, 1], [1,...  \n",
       "...                                                 ...  \n",
       "2393  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [0,...  \n",
       "2394  [[0, 0, 0, 1], [0, 0, 1, 0], [1, 0, 0, 0], [0,...  \n",
       "2395  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...  \n",
       "2396  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [1,...  \n",
       "2397  [[0, 0, 1, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "\n",
       "[2398 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     HOM04D006130  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...            ...                                                ...   \n",
       "2789  HOM04D004133  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2791  HOM04D006357  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2789  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2791  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2794 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL1G44210</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G30260</td>\n",
       "      <td>HOM04D006130</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2789</th>\n",
       "      <td>Carub.0004s0553</td>\n",
       "      <td>0</td>\n",
       "      <td>AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25740</td>\n",
       "      <td>HOM04D004133</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2791</th>\n",
       "      <td>Carub.0005s2520</td>\n",
       "      <td>0</td>\n",
       "      <td>GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58460</td>\n",
       "      <td>HOM04D006357</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2794 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "4           AL1G44210     1   \n",
       "...               ...   ...   \n",
       "2789  Carub.0004s0553     0   \n",
       "2790  Carub.0006s3101     1   \n",
       "2791  Carub.0005s2520     0   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "4     CGGCAATGTAATTAATACCAAAACCAAATGGGCGGCACTACCCGCT...     aly  AT1G30260   \n",
       "...                                                 ...     ...        ...   \n",
       "2789  AATCCAACGGCATGAGTACATGGGTTTCGAGAACCTTTGCAAGTAC...     cru  AT2G25740   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2791  GCAGCTGAACTATTGGGTGTTCGTCCTTGTGTCTTTAAGCCTCTTC...     cru  AT3G58460   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "4     HOM04D006130  [[0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 1, 0], [0,...   \n",
       "...            ...                                                ...   \n",
       "2789  HOM04D004133  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2791  HOM04D006357  [[0, 0, 1, 0], [0, 1, 0, 0], [1, 0, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "4     [[0, 0, 0, 1], [0, 0, 1, 0], [0, 0, 1, 0], [1,...  \n",
       "...                                                 ...  \n",
       "2789  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2791  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2794 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylo_aug_df = phylo_aug_df[phylo_aug_df[\"GeneFamily\"].notna()]\n",
    "phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Train-test split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.prepare_dataset import grouped_shuffle_split\n",
    "train_df, validation_test_df = grouped_shuffle_split(dataset, dataset[\"GeneFamily\"], 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "validation_df, test_df  = grouped_shuffle_split(validation_test_df, validation_test_df[\"GeneFamily\"], 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = train_df.sample(700, random_state=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of training set: 700\n",
      "Length of validation set: 257\n",
      "Length of test set: 241\n"
     ]
    }
   ],
   "source": [
    "print(f\"Length of training set: {train_df.shape[0]}\")\n",
    "print(f\"Length of validation set: {validation_df.shape[0]}\")\n",
    "print(f\"Length of test set: {test_df.shape[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out the orthologs from samples in test or validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_phylo_aug_df = phylo_aug_df[~phylo_aug_df[\"Ortholog\"].isin(validation_test_df[\"GeneID\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Label</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Species</th>\n",
       "      <th>Ortholog</th>\n",
       "      <th>GeneFamily</th>\n",
       "      <th>One_hot_encoded</th>\n",
       "      <th>RC_one_hot_encoded</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AL1G34890</td>\n",
       "      <td>0</td>\n",
       "      <td>AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT1G21720</td>\n",
       "      <td>HOM04D003099</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AL6G48600</td>\n",
       "      <td>1</td>\n",
       "      <td>CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT4G05010</td>\n",
       "      <td>HOM04D003060</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AL6G13960</td>\n",
       "      <td>1</td>\n",
       "      <td>TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G04340</td>\n",
       "      <td>HOM04D000135</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AL4G45030</td>\n",
       "      <td>1</td>\n",
       "      <td>ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT2G45820</td>\n",
       "      <td>HOM04D001302</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AL6G25970</td>\n",
       "      <td>0</td>\n",
       "      <td>GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...</td>\n",
       "      <td>aly</td>\n",
       "      <td>AT5G15150</td>\n",
       "      <td>HOM04D000209</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2787</th>\n",
       "      <td>Carub.0005s1225</td>\n",
       "      <td>1</td>\n",
       "      <td>GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G46450</td>\n",
       "      <td>HOM04D003906</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2788</th>\n",
       "      <td>Carub.0005s2556</td>\n",
       "      <td>1</td>\n",
       "      <td>TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58780</td>\n",
       "      <td>HOM04D000037</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2790</th>\n",
       "      <td>Carub.0006s3101</td>\n",
       "      <td>1</td>\n",
       "      <td>TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT4G11910</td>\n",
       "      <td>HOM04D001924</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...</td>\n",
       "      <td>[[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2792</th>\n",
       "      <td>Carub.0005s2528</td>\n",
       "      <td>0</td>\n",
       "      <td>GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT3G58530</td>\n",
       "      <td>HOM04D007279</td>\n",
       "      <td>[[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2793</th>\n",
       "      <td>Carub.0004s0582</td>\n",
       "      <td>0</td>\n",
       "      <td>TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...</td>\n",
       "      <td>cru</td>\n",
       "      <td>AT2G25970</td>\n",
       "      <td>HOM04D001540</td>\n",
       "      <td>[[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...</td>\n",
       "      <td>[[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2217 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GeneID Label  \\\n",
       "0           AL1G34890     0   \n",
       "1           AL6G48600     1   \n",
       "2           AL6G13960     1   \n",
       "3           AL4G45030     1   \n",
       "5           AL6G25970     0   \n",
       "...               ...   ...   \n",
       "2787  Carub.0005s1225     1   \n",
       "2788  Carub.0005s2556     1   \n",
       "2790  Carub.0006s3101     1   \n",
       "2792  Carub.0005s2528     0   \n",
       "2793  Carub.0004s0582     0   \n",
       "\n",
       "                                               Sequence Species   Ortholog  \\\n",
       "0     AACTGACACCTAAACTCCACTGCAGAGTGGCAGAAGCATTTGTAAG...     aly  AT1G21720   \n",
       "1     CCTACGCCACATTAACCTGATCATACCTAAACACGGTGAATCATCA...     aly  AT4G05010   \n",
       "2     TTAGTCGATGCTTTATTGAAAAATATATAGTTTGTAAAACAGTCAT...     aly  AT5G04340   \n",
       "3     ATCGATGTAGATATTATCTTATTCTTATACATTTAATACAGATTCA...     aly  AT2G45820   \n",
       "5     GATTGAGAGTGGGTGTCATAGTTTCTTTTAATTTTTTTGCTTATCT...     aly  AT5G15150   \n",
       "...                                                 ...     ...        ...   \n",
       "2787  GAAAGAATGGAGAGAGAAGAGAGACGAGTTCAAGAAGAAAGTGAGC...     cru  AT3G46450   \n",
       "2788  TTGTGTAGTGCCTTGAAAACGTCCTGATCCCTCCAAAGATGTTACA...     cru  AT3G58780   \n",
       "2790  TTCTTTCTCTTTCTCACTATAATAAATATATAATGAAAAAAAAATT...     cru  AT4G11910   \n",
       "2792  GCGCAAGAAGCGGAGAACAGGGACGGTGAGACCGAGGAGTTTTTGC...     cru  AT3G58530   \n",
       "2793  TTGTCGTCCAATAGCATATGCACACCTTTTGTACTCTTATATTATT...     cru  AT2G25970   \n",
       "\n",
       "        GeneFamily                                    One_hot_encoded  \\\n",
       "0     HOM04D003099  [[1, 0, 0, 0], [1, 0, 0, 0], [0, 1, 0, 0], [0,...   \n",
       "1     HOM04D003060  [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [1,...   \n",
       "2     HOM04D000135  [[0, 0, 0, 1], [0, 0, 0, 1], [1, 0, 0, 0], [0,...   \n",
       "3     HOM04D001302  [[1, 0, 0, 0], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "5     HOM04D000209  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...   \n",
       "...            ...                                                ...   \n",
       "2787  HOM04D003906  [[0, 0, 1, 0], [1, 0, 0, 0], [1, 0, 0, 0], [1,...   \n",
       "2788  HOM04D000037  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "2790  HOM04D001924  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 1, 0, 0], [0,...   \n",
       "2792  HOM04D007279  [[0, 0, 1, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0,...   \n",
       "2793  HOM04D001540  [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 1, 0], [0,...   \n",
       "\n",
       "                                     RC_one_hot_encoded  \n",
       "0     [[0, 1, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "1     [[1, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1], [1,...  \n",
       "2     [[1, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "3     [[0, 0, 0, 1], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "5     [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 0, 1], [0,...  \n",
       "...                                                 ...  \n",
       "2787  [[1, 0, 0, 0], [0, 0, 0, 1], [1, 0, 0, 0], [0,...  \n",
       "2788  [[0, 0, 0, 1], [1, 0, 0, 0], [1, 0, 0, 0], [1,...  \n",
       "2790  [[1, 0, 0, 0], [0, 0, 1, 0], [0, 1, 0, 0], [0,...  \n",
       "2792  [[0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 0, 1], [0,...  \n",
       "2793  [[0, 0, 1, 0], [1, 0, 0, 0], [0, 0, 1, 0], [0,...  \n",
       "\n",
       "[2217 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_phylo_aug_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch time augmentation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "\n",
    "def get_batch(data, phylo_aug_data, indices, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tCreates a batch of the input and one-hot encodes the sequences\n",
    "\t\"\"\"\n",
    "\tseqs = []\n",
    "\tlabels = []\n",
    "\tgene_ids = []\n",
    "\tfor ii in indices:\n",
    "\t\tgene_id = data.iloc[ii].GeneID\n",
    "\t\tsample = data.iloc[ii]\n",
    "\t\tif random.choices(population=[0, 1], weights=[1 - phylo_aug_rate, phylo_aug_rate])[0] == 1:\n",
    "\t\t\torthologs = phylo_aug_data[phylo_aug_data.Ortholog == gene_id]\n",
    "\t\t\tif not orthologs.empty:\n",
    "\t\t\t\tsample = phylo_aug_data[phylo_aug_data.Ortholog == gene_id].sample().iloc[0]\n",
    "\t\t\t# Randomly pick the forward or reverse complement strand\n",
    "\t\tif random.randint(0, 1) == 0:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.RC_one_hot_encoded)\n",
    "\t\telse:\n",
    "\t\t\tgene_ids.append(sample.GeneID)\n",
    "\t\t\tseqs.append(sample.One_hot_encoded)\n",
    "\t\t\t\n",
    "\t\tlabels.append(sample.Label)\n",
    "\n",
    "\tX_batch = np.array(seqs)\n",
    "\tX_batch = X_batch.astype('float32')\n",
    "\n",
    "\tY_batch = np.array(labels)\n",
    "\tY_batch = Y_batch.astype('int64')\n",
    "\n",
    "\treturn X_batch, Y_batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_gen(data, phylo_aug_data, batch_size, perform_phylo_aug=True, phylo_aug_rate=1.0):\n",
    "\t\"\"\"\n",
    "\tGenerator function for loading input data in batches\n",
    "\t\"\"\"\n",
    "\tnum_samples = data.shape[0]\n",
    "\t# indices = list(range(num_samples))\n",
    "\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t# indices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\n",
    "\tii = 0\n",
    "\twhile True:\n",
    "\t\tyield get_batch(data, phylo_aug_data, indices[ii:ii + batch_size], perform_phylo_aug, phylo_aug_rate)\n",
    "\t\tii += batch_size\n",
    "\t\tif ii >= num_samples:\n",
    "\t\t\tii = 0\n",
    "\t\t\tindices = np.random.choice(list(range(num_samples)), num_samples, replace=False)\n",
    "\t\t\t# indices = list(range(num_samples))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare validation input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_input_and_labels(df):\n",
    "\tohe_np = np.stack(df[\"One_hot_encoded\"])\n",
    "\trc_np = np.stack(df[\"RC_one_hot_encoded\"])\n",
    "\n",
    "\tx = np.append(ohe_np, rc_np, axis=0)\n",
    "\tx = x.astype('float32')\n",
    "\ty = np.append(df[\"Label\"], df[\"Label\"])\n",
    "\ty = y.astype('int64')\n",
    "\treturn x, y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_validation, y_validation = get_input_and_labels(validation_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we do batch time augmentation we define the data generator here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "BATCH_SIZE = 100\n",
    "train_data_gen = data_gen(train_df,filtered_phylo_aug_df, BATCH_SIZE, True, 0.6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the shape of the input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_shape = train_df[\"One_hot_encoded\"].iloc[0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Initialize the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = get_model(input_shape=input_shape, perform_evoaug=False,learning_rate=0.001)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# early stopping callback\n",
    "import tensorflow as tf\n",
    "\n",
    "early_stopping_callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\tpatience=40,\n",
    "\t\t\t\t\t\t\t\t\t\t\tverbose=1,\n",
    "\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\trestore_best_weights=True)\n",
    "# reduce learning rate callback\n",
    "reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(monitor='val_loss',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tfactor=0.1,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tpatience=5,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmin_lr=1e-7,\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tmode='min',\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tverbose=1)\n",
    "callbacks = [early_stopping_callback, reduce_lr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "7/7 [==============================] - 3s 208ms/step - loss: 0.6995 - acc: 0.4643 - auROC: 0.4453 - auPRC: 0.4597 - true_positives_1: 274.0000 - val_loss: 0.6937 - val_acc: 0.4961 - val_auROC: 0.4935 - val_auPRC: 0.4895 - val_true_positives_1: 247.0000 - lr: 0.0010\n",
      "Epoch 2/100\n",
      "7/7 [==============================] - 1s 104ms/step - loss: 0.6954 - acc: 0.4800 - auROC: 0.4862 - auPRC: 0.4895 - true_positives_1: 66.0000 - val_loss: 0.6946 - val_acc: 0.4981 - val_auROC: 0.5046 - val_auPRC: 0.5042 - val_true_positives_1: 0.0000e+00 - lr: 0.0010\n",
      "Epoch 3/100\n",
      "7/7 [==============================] - 1s 105ms/step - loss: 0.6945 - acc: 0.4971 - auROC: 0.5066 - auPRC: 0.5060 - true_positives_1: 27.0000 - val_loss: 0.6936 - val_acc: 0.4864 - val_auROC: 0.4841 - val_auPRC: 0.4889 - val_true_positives_1: 29.0000 - lr: 0.0010\n",
      "Epoch 4/100\n",
      "7/7 [==============================] - 1s 102ms/step - loss: 0.6939 - acc: 0.5071 - auROC: 0.4876 - auPRC: 0.4853 - true_positives_1: 239.0000 - val_loss: 0.6937 - val_acc: 0.5019 - val_auROC: 0.4838 - val_auPRC: 0.4848 - val_true_positives_1: 258.0000 - lr: 0.0010\n",
      "Epoch 5/100\n",
      "7/7 [==============================] - 1s 107ms/step - loss: 0.6928 - acc: 0.5200 - auROC: 0.5405 - auPRC: 0.5275 - true_positives_1: 292.0000 - val_loss: 0.6937 - val_acc: 0.5000 - val_auROC: 0.4803 - val_auPRC: 0.4893 - val_true_positives_1: 255.0000 - lr: 0.0010\n",
      "Epoch 6/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6932 - acc: 0.5071 - auROC: 0.5107 - auPRC: 0.5157 - true_positives_1: 223.0000\n",
      "Epoch 6: ReduceLROnPlateau reducing learning rate to 0.00010000000474974513.\n",
      "7/7 [==============================] - 1s 107ms/step - loss: 0.6932 - acc: 0.5071 - auROC: 0.5107 - auPRC: 0.5157 - true_positives_1: 223.0000 - val_loss: 0.6939 - val_acc: 0.4630 - val_auROC: 0.4722 - val_auPRC: 0.4743 - val_true_positives_1: 117.0000 - lr: 0.0010\n",
      "Epoch 7/100\n",
      "7/7 [==============================] - 1s 138ms/step - loss: 0.6928 - acc: 0.5114 - auROC: 0.5196 - auPRC: 0.5042 - true_positives_1: 194.0000 - val_loss: 0.6939 - val_acc: 0.4591 - val_auROC: 0.4743 - val_auPRC: 0.4726 - val_true_positives_1: 96.0000 - lr: 1.0000e-04\n",
      "Epoch 8/100\n",
      "7/7 [==============================] - 1s 119ms/step - loss: 0.6921 - acc: 0.5329 - auROC: 0.5503 - auPRC: 0.5320 - true_positives_1: 165.0000 - val_loss: 0.6939 - val_acc: 0.4747 - val_auROC: 0.4777 - val_auPRC: 0.4919 - val_true_positives_1: 60.0000 - lr: 1.0000e-04\n",
      "Epoch 9/100\n",
      "7/7 [==============================] - 1s 114ms/step - loss: 0.6937 - acc: 0.4886 - auROC: 0.4968 - auPRC: 0.5101 - true_positives_1: 168.0000 - val_loss: 0.6939 - val_acc: 0.4650 - val_auROC: 0.4777 - val_auPRC: 0.4917 - val_true_positives_1: 51.0000 - lr: 1.0000e-04\n",
      "Epoch 10/100\n",
      "7/7 [==============================] - 1s 107ms/step - loss: 0.6927 - acc: 0.5500 - auROC: 0.5357 - auPRC: 0.5265 - true_positives_1: 189.0000 - val_loss: 0.6939 - val_acc: 0.4767 - val_auROC: 0.4771 - val_auPRC: 0.4769 - val_true_positives_1: 68.0000 - lr: 1.0000e-04\n",
      "Epoch 11/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6935 - acc: 0.5143 - auROC: 0.5125 - auPRC: 0.5081 - true_positives_1: 163.0000\n",
      "Epoch 11: ReduceLROnPlateau reducing learning rate to 1.0000000474974514e-05.\n",
      "7/7 [==============================] - 1s 109ms/step - loss: 0.6935 - acc: 0.5143 - auROC: 0.5125 - auPRC: 0.5081 - true_positives_1: 163.0000 - val_loss: 0.6939 - val_acc: 0.4747 - val_auROC: 0.4792 - val_auPRC: 0.4792 - val_true_positives_1: 61.0000 - lr: 1.0000e-04\n",
      "Epoch 12/100\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6928 - acc: 0.5271 - auROC: 0.5233 - auPRC: 0.5110 - true_positives_1: 183.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4786 - val_auPRC: 0.4800 - val_true_positives_1: 58.0000 - lr: 1.0000e-05\n",
      "Epoch 13/100\n",
      "7/7 [==============================] - 1s 109ms/step - loss: 0.6930 - acc: 0.5229 - auROC: 0.5241 - auPRC: 0.5104 - true_positives_1: 166.0000 - val_loss: 0.6939 - val_acc: 0.4747 - val_auROC: 0.4768 - val_auPRC: 0.4754 - val_true_positives_1: 60.0000 - lr: 1.0000e-05\n",
      "Epoch 14/100\n",
      "7/7 [==============================] - 1s 108ms/step - loss: 0.6917 - acc: 0.5243 - auROC: 0.5542 - auPRC: 0.5382 - true_positives_1: 185.0000 - val_loss: 0.6939 - val_acc: 0.4767 - val_auROC: 0.4798 - val_auPRC: 0.4785 - val_true_positives_1: 61.0000 - lr: 1.0000e-05\n",
      "Epoch 15/100\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6942 - acc: 0.4843 - auROC: 0.4889 - auPRC: 0.5014 - true_positives_1: 163.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4756 - val_auPRC: 0.4776 - val_true_positives_1: 59.0000 - lr: 1.0000e-05\n",
      "Epoch 16/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6933 - acc: 0.5143 - auROC: 0.5196 - auPRC: 0.5037 - true_positives_1: 173.0000\n",
      "Epoch 16: ReduceLROnPlateau reducing learning rate to 1.0000000656873453e-06.\n",
      "7/7 [==============================] - 1s 109ms/step - loss: 0.6933 - acc: 0.5143 - auROC: 0.5196 - auPRC: 0.5037 - true_positives_1: 173.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4726 - val_auPRC: 0.4748 - val_true_positives_1: 59.0000 - lr: 1.0000e-05\n",
      "Epoch 17/100\n",
      "7/7 [==============================] - 1s 115ms/step - loss: 0.6940 - acc: 0.5029 - auROC: 0.4974 - auPRC: 0.5214 - true_positives_1: 175.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-06\n",
      "Epoch 18/100\n",
      "7/7 [==============================] - 1s 105ms/step - loss: 0.6920 - acc: 0.5229 - auROC: 0.5385 - auPRC: 0.5380 - true_positives_1: 174.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-06\n",
      "Epoch 19/100\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6915 - acc: 0.5486 - auROC: 0.5728 - auPRC: 0.5504 - true_positives_1: 184.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-06\n",
      "Epoch 20/100\n",
      "7/7 [==============================] - 1s 103ms/step - loss: 0.6930 - acc: 0.5100 - auROC: 0.5131 - auPRC: 0.5161 - true_positives_1: 165.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-06\n",
      "Epoch 21/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6938 - acc: 0.5129 - auROC: 0.5035 - auPRC: 0.4855 - true_positives_1: 168.0000\n",
      "Epoch 21: ReduceLROnPlateau reducing learning rate to 1.0000001111620805e-07.\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6938 - acc: 0.5129 - auROC: 0.5035 - auPRC: 0.4855 - true_positives_1: 168.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-06\n",
      "Epoch 22/100\n",
      "7/7 [==============================] - 1s 107ms/step - loss: 0.6940 - acc: 0.5057 - auROC: 0.4868 - auPRC: 0.4986 - true_positives_1: 168.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 23/100\n",
      "7/7 [==============================] - 1s 107ms/step - loss: 0.6937 - acc: 0.4886 - auROC: 0.4922 - auPRC: 0.5107 - true_positives_1: 158.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 24/100\n",
      "7/7 [==============================] - 1s 104ms/step - loss: 0.6932 - acc: 0.5157 - auROC: 0.5134 - auPRC: 0.4999 - true_positives_1: 168.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 25/100\n",
      "7/7 [==============================] - 1s 105ms/step - loss: 0.6927 - acc: 0.5543 - auROC: 0.5320 - auPRC: 0.5184 - true_positives_1: 183.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 26/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6938 - acc: 0.5286 - auROC: 0.5010 - auPRC: 0.4989 - true_positives_1: 175.0000\n",
      "Epoch 26: ReduceLROnPlateau reducing learning rate to 1e-07.\n",
      "7/7 [==============================] - 1s 103ms/step - loss: 0.6938 - acc: 0.5286 - auROC: 0.5010 - auPRC: 0.4989 - true_positives_1: 175.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 27/100\n",
      "7/7 [==============================] - 1s 101ms/step - loss: 0.6934 - acc: 0.4914 - auROC: 0.5006 - auPRC: 0.5038 - true_positives_1: 167.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 28/100\n",
      "7/7 [==============================] - 1s 105ms/step - loss: 0.6929 - acc: 0.5071 - auROC: 0.5190 - auPRC: 0.5109 - true_positives_1: 175.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 29/100\n",
      "7/7 [==============================] - 1s 105ms/step - loss: 0.6920 - acc: 0.5257 - auROC: 0.5506 - auPRC: 0.5469 - true_positives_1: 169.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 30/100\n",
      "7/7 [==============================] - 1s 101ms/step - loss: 0.6943 - acc: 0.4800 - auROC: 0.4756 - auPRC: 0.4942 - true_positives_1: 163.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 31/100\n",
      "7/7 [==============================] - 1s 101ms/step - loss: 0.6948 - acc: 0.4529 - auROC: 0.4748 - auPRC: 0.4843 - true_positives_1: 149.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 32/100\n",
      "7/7 [==============================] - 1s 101ms/step - loss: 0.6932 - acc: 0.5014 - auROC: 0.5178 - auPRC: 0.5077 - true_positives_1: 173.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 33/100\n",
      "7/7 [==============================] - 1s 100ms/step - loss: 0.6935 - acc: 0.5129 - auROC: 0.5096 - auPRC: 0.5191 - true_positives_1: 164.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 34/100\n",
      "7/7 [==============================] - 1s 101ms/step - loss: 0.6947 - acc: 0.4914 - auROC: 0.4692 - auPRC: 0.4852 - true_positives_1: 164.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 35/100\n",
      "7/7 [==============================] - 1s 111ms/step - loss: 0.6919 - acc: 0.5371 - auROC: 0.5549 - auPRC: 0.5515 - true_positives_1: 184.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 36/100\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6928 - acc: 0.5000 - auROC: 0.5219 - auPRC: 0.5265 - true_positives_1: 165.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 37/100\n",
      "7/7 [==============================] - 1s 104ms/step - loss: 0.6941 - acc: 0.4729 - auROC: 0.4835 - auPRC: 0.4828 - true_positives_1: 151.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 38/100\n",
      "7/7 [==============================] - 1s 103ms/step - loss: 0.6942 - acc: 0.4814 - auROC: 0.4834 - auPRC: 0.4853 - true_positives_1: 170.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 39/100\n",
      "7/7 [==============================] - 1s 102ms/step - loss: 0.6928 - acc: 0.5029 - auROC: 0.5321 - auPRC: 0.5206 - true_positives_1: 167.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 40/100\n",
      "7/7 [==============================] - 1s 115ms/step - loss: 0.6928 - acc: 0.5086 - auROC: 0.5232 - auPRC: 0.5030 - true_positives_1: 172.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 41/100\n",
      "7/7 [==============================] - 1s 104ms/step - loss: 0.6929 - acc: 0.5000 - auROC: 0.5213 - auPRC: 0.5231 - true_positives_1: 166.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 42/100\n",
      "7/7 [==============================] - 1s 103ms/step - loss: 0.6920 - acc: 0.5414 - auROC: 0.5514 - auPRC: 0.5374 - true_positives_1: 183.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 43/100\n",
      "7/7 [==============================] - ETA: 0s - loss: 0.6921 - acc: 0.5229 - auROC: 0.5403 - auPRC: 0.5293 - true_positives_1: 178.0000Restoring model weights from the end of the best epoch: 3.\n",
      "7/7 [==============================] - 1s 106ms/step - loss: 0.6921 - acc: 0.5229 - auROC: 0.5403 - auPRC: 0.5293 - true_positives_1: 178.0000 - val_loss: 0.6939 - val_acc: 0.4728 - val_auROC: 0.4744 - val_auPRC: 0.4757 - val_true_positives_1: 59.0000 - lr: 1.0000e-07\n",
      "Epoch 43: early stopping\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "history = model.fit(train_data_gen,\n",
    "\t\t\t\t\tsteps_per_epoch=math.ceil(\n",
    "\t\t\t\t\t\ttrain_df.shape[0] / BATCH_SIZE),\n",
    "\t\t\t\t\tepochs=100,\n",
    "\t\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\t\tcallbacks=callbacks\n",
    "\t\t\t\t\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'loss': 0.6945062875747681,\n",
       " 'acc': 0.49714285135269165,\n",
       " 'auROC': 0.5066490173339844,\n",
       " 'auPRC': 0.5060290098190308,\n",
       " 'true_positives_1': 27.0,\n",
       " 'val_loss': 0.6936250925064087,\n",
       " 'val_acc': 0.48638132214546204,\n",
       " 'val_auROC': 0.4841478765010834,\n",
       " 'val_auPRC': 0.48889878392219543,\n",
       " 'val_true_positives_1': 29.0,\n",
       " 'lr': 0.001}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auroc_values = history.history['val_loss']\n",
    "\n",
    "# Find the index of the epoch where AUROC is the maximum\n",
    "max_auroc_epoch = np.argmin(auroc_values)\n",
    "\n",
    "# Extract all the metrics for the epoch with the maximum AUROC\n",
    "metrics_at_max_auroc = {metric: values[max_auroc_epoch] for metric, values in history.history.items()}\n",
    "metrics_at_max_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Finetune phylo-aug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_train_pure, y_train_pure = get_input_and_labels(train_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/10\n",
      "14/14 [==============================] - 2s 40ms/step - loss: 0.6491 - acc: 0.6329 - auROC: 0.6704 - auPRC: 0.6425 - true_positives_1: 454.0000 - val_loss: 0.6621 - val_acc: 0.6051 - val_auROC: 0.6491 - val_auPRC: 0.6293 - val_true_positives_1: 202.0000\n",
      "Epoch 2/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6437 - acc: 0.6343 - auROC: 0.6805 - auPRC: 0.6409 - true_positives_1: 482.0000 - val_loss: 0.6576 - val_acc: 0.6109 - val_auROC: 0.6508 - val_auPRC: 0.6300 - val_true_positives_1: 165.0000\n",
      "Epoch 3/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6294 - acc: 0.6514 - auROC: 0.7026 - auPRC: 0.6764 - true_positives_1: 458.0000 - val_loss: 0.6587 - val_acc: 0.6128 - val_auROC: 0.6503 - val_auPRC: 0.6297 - val_true_positives_1: 185.0000\n",
      "Epoch 4/10\n",
      "14/14 [==============================] - 0s 18ms/step - loss: 0.6442 - acc: 0.6386 - auROC: 0.6813 - auPRC: 0.6670 - true_positives_1: 486.0000 - val_loss: 0.6591 - val_acc: 0.6051 - val_auROC: 0.6507 - val_auPRC: 0.6320 - val_true_positives_1: 187.0000\n",
      "Epoch 5/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6365 - acc: 0.6364 - auROC: 0.6918 - auPRC: 0.6786 - true_positives_1: 442.0000 - val_loss: 0.6580 - val_acc: 0.6128 - val_auROC: 0.6503 - val_auPRC: 0.6300 - val_true_positives_1: 181.0000\n",
      "Epoch 6/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6361 - acc: 0.6336 - auROC: 0.6918 - auPRC: 0.6693 - true_positives_1: 480.0000 - val_loss: 0.6580 - val_acc: 0.6109 - val_auROC: 0.6511 - val_auPRC: 0.6299 - val_true_positives_1: 181.0000\n",
      "Epoch 7/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6276 - acc: 0.6543 - auROC: 0.7050 - auPRC: 0.6822 - true_positives_1: 463.0000 - val_loss: 0.6589 - val_acc: 0.6031 - val_auROC: 0.6505 - val_auPRC: 0.6305 - val_true_positives_1: 184.0000\n",
      "Epoch 8/10\n",
      "14/14 [==============================] - 0s 18ms/step - loss: 0.6332 - acc: 0.6421 - auROC: 0.6944 - auPRC: 0.6765 - true_positives_1: 501.0000 - val_loss: 0.6591 - val_acc: 0.6031 - val_auROC: 0.6501 - val_auPRC: 0.6306 - val_true_positives_1: 182.0000\n",
      "Epoch 9/10\n",
      "14/14 [==============================] - 0s 18ms/step - loss: 0.6318 - acc: 0.6371 - auROC: 0.6960 - auPRC: 0.6760 - true_positives_1: 446.0000 - val_loss: 0.6592 - val_acc: 0.6128 - val_auROC: 0.6496 - val_auPRC: 0.6295 - val_true_positives_1: 179.0000\n",
      "Epoch 10/10\n",
      "14/14 [==============================] - 0s 19ms/step - loss: 0.6354 - acc: 0.6457 - auROC: 0.6943 - auPRC: 0.6644 - true_positives_1: 514.0000 - val_loss: 0.6609 - val_acc: 0.5973 - val_auROC: 0.6474 - val_auPRC: 0.6236 - val_true_positives_1: 187.0000\n"
     ]
    }
   ],
   "source": [
    "from tensorflow import keras\n",
    "from src.diff_expression_model import compile_model\n",
    "\n",
    "# set up callbacks\n",
    "model = compile_model(model, learning_rate=0.0001)\n",
    "\n",
    "# train model\n",
    "finetune_history = model.fit(x_train_pure, y_train_pure,\n",
    "\t\t\t\t\t\tepochs=10,\n",
    "\t\t\t\t\t\tbatch_size=100,\n",
    "\t\t\t\t\t\tshuffle=True,\n",
    "\t\t\t\t\t\tvalidation_data=(x_validation, y_validation),\n",
    "\t\t\t\t\t\tcallbacks=[early_stopping_callback])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'loss': 0.6436826586723328,\n",
       " 'acc': 0.6342856884002686,\n",
       " 'auROC': 0.6804790496826172,\n",
       " 'auPRC': 0.6408889889717102,\n",
       " 'true_positives_1': 482.0,\n",
       " 'val_loss': 0.6576040387153625,\n",
       " 'val_acc': 0.6108949184417725,\n",
       " 'val_auROC': 0.6508070230484009,\n",
       " 'val_auPRC': 0.6299607157707214,\n",
       " 'val_true_positives_1': 165.0}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auroc_values = finetune_history.history['val_loss']\n",
    "\n",
    "# Find the index of the epoch where AUROC is the maximum\n",
    "max_auroc_epoch = np.argmin(auroc_values)\n",
    "\n",
    "# Extract all the metrics for the epoch with the maximum AUROC\n",
    "metrics_at_max_auroc = {metric: values[max_auroc_epoch] for metric, values in finetune_history.history.items()}\n",
    "metrics_at_max_auroc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test set"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We perform post-hoc conjoing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "siamese_model = get_siamese_model(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare the test set data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_test = np.stack(test_df[\"One_hot_encoded\"])\n",
    "x_test_rc = np.stack(test_df[\"RC_one_hot_encoded\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_test = test_df[\"Label\"].to_numpy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we used evo aug in our model all the sequences which were trained on, were of length 620. We use the evoaug padding function to pad the test set to the same length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Sequential' object has no attribute '_pad_end'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[44], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m x_test \u001b[38;5;241m=\u001b[39m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_pad_end\u001b[49m(x_test)\n\u001b[1;32m      2\u001b[0m x_test_rc \u001b[38;5;241m=\u001b[39m model\u001b[38;5;241m.\u001b[39m_pad_end(x_test_rc)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Sequential' object has no attribute '_pad_end'"
     ]
    }
   ],
   "source": [
    "x_test = model._pad_end(x_test)\n",
    "x_test_rc = model._pad_end(x_test_rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8/8 [==============================] - 0s 4ms/step\n"
     ]
    }
   ],
   "source": [
    "predictions_categories, predictions = post_hoc_conjoining(siamese_model, x_test, x_test_rc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6568953568953568"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_auroc(y_test, predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.64      0.64      0.64       111\n",
      "           1       0.69      0.69      0.69       130\n",
      "\n",
      "    accuracy                           0.67       241\n",
      "   macro avg       0.67      0.67      0.67       241\n",
      "weighted avg       0.67      0.67      0.67       241\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(y_test, predictions_categories))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
